---
title: 'Tampere PC: Gene expression vs ATAC signal around TSS corrlation (all genes and DEG)'
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Francesco Tabaro"
output:
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: true  
    df_print: tibble
    fig_caption: true
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
    fig_caption: true
    df_print: paged
    toc_float:
      collapsed: false
      smooth_scroll: false
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir="/bmt-data/genomics//projects//atac_workdir/")
```

```{r, include=FALSE, echo=FALSE}
library(data.table)
library(DESeq2)
library(gridExtra)
library(matrixStats)
library(ggplot2)
```

# Load data

## Annotation

Load Ensembl v90 GTF annotation.

```{r}
source("R/getAnnotationTable.R")
ann <- getAnnotationTable()
```

## RNA

```{r}
load("R_data/RNAdds.RData")
load("R_data/RNA_differentiallyExpressed.RData")
```

## ATAC

Load ATAC-seq signal at TSS of quantified genes.

```{r}
ATAC_MAT_PATH    <- "peak_tss_dar_quantifications/tss_quantification.tsv.gz"

atac_rc_mat     <- fread(cmd = sprintf("zcat %s", ATAC_MAT_PATH), sep ="\t", header = TRUE, showProgress = FALSE)
prom_gid        <- fread("tss_grid/all_genes_500u500d_sorted.bed")
names(prom_gid) <- c("chrom", "start", "end", "gid", "score", "strand")

atac_rc_mat     <- merge(prom_gid, atac_rc_mat, by=c("chrom", "start", "end"), all.x=TRUE)
atac_rc_mat[,`:=`(score=NULL,strand=NULL)]
setkey(atac_rc_mat, gid)

# atac_rc_mat[,`:=`(dar_id=paste(chrom, start, end, sep="_"))]
# setkey(atac_rc_mat, dar_id)
```


## Find common set of samples

```{r}
columns <- intersect(names(atac_rc_mat), colnames(rna_dds))

shared.samples <- list(bph = columns[grep("^BPH",columns)],
                       pc = columns[grep("^PC",columns)],
                       crpc = columns[grep("^CRPC",columns)])

atac.samples <- list(bph = names(atac_rc_mat)[grepl("^BPH",names(atac_rc_mat))], 
                     pc = names(atac_rc_mat)[grepl("^PC",names(atac_rc_mat))], 
                     crpc = names(atac_rc_mat)[grepl("^CRPC",names(atac_rc_mat))])

rna.samples <- list(bph = colnames(rna_dds)[grepl("^BPH",colnames(rna_dds))], 
                     pc = colnames(rna_dds)[grepl("^PC",colnames(rna_dds))], 
                     crpc = colnames(rna_dds)[grepl("^CRPC",colnames(rna_dds))])
```


# Compute log2 fold change

```{r}
computeLFC <- function (rc, sample.set1, sample.set2) {
  log2(median(rc[sample.set1]) + 1) - log2(median(rc[sample.set2]) + 1)
}
```


# Define helper functions to compute correlation coefficients

Correlation will be computed between normalized RNA-seq signal of genes and normalized ATAC-seq signal quantified in proximal promoter region defined from -500bp to +500bp around TSS.

```{r}
computeCor <- function (ensgid, atac_rc, method, samples) {
  rna_rc <- assay(rna_dds,"norm.counts")[ensgid, match(samples,colnames(rna_dds))]
  atac_rc <- atac_rc[match(samples, names(atac_rc))]
  stopifnot(all(names(atac_rc) == names(rna_rc)))
  cor(atac_rc, rna_rc, method=method, use="na.or.complete")
}

getQuadrant <- function (rna_lfc, atac_lfc) {
  apply(cbind(atac_lfc, rna_lfc),1,function(row){
    quadrant <- NA
    if(row[1] >= 0 & row[2] >= 0) {
      quadrant <- 1
    } else if (row[1] >= 0 & row[2] < 0) {
      quadrant <- 2
    } else if (row[1] < 0 & row[2] < 0) {
      quadrant <- 3
    } else if (row[1] < 0 & row[2] >= 0) {
      quadrant <- 4
    }
    return(quadrant)
  })
}

doCor <- function (target, ncores = parallel::detectCores()) {
  
  lfc <- vector("list", length = length(target))
  names(lfc) <- names(target)
  
  for (name in names(target)) {
    
    x <- target[[name]]
    
    de <- x[[1]]
    geneset <- rownames(de)
    
    s1 <- atac.samples[[tolower(x[[2]][1])]]
    s2 <- atac.samples[[tolower(x[[2]][2])]]
    sele_cols <- c(s1,s2)
    
    # s <- c(shared.samples[[tolower(x[[2]][1])]],
    #        shared.samples[[tolower(x[[2]][2])]])
    s <- Reduce(c, shared.samples) # always use all samples to compute correlation
    
    res <- mclapply(seq_along(geneset), function (i) {

      gene    <- geneset[i]
      
      tss_rc  <- setNames(as.numeric(atac_rc_mat[gid == gene, ..sele_cols]), sele_cols)
      lfc     <- computeLFC(tss_rc, s1, s2)
      
      tss_rc  <- setNames(as.numeric(atac_rc_mat[gid == gene, ..s]), s)
      pearson  <- computeCor(gene, tss_rc, "pearson", s)
      spearman <- computeCor(gene, tss_rc, "spearman", s)

      return(data.frame(row.names=gene, 
                        atac_log2FoldChange=lfc,
                        pearson=pearson,
                        spearman=spearman))
    }, mc.cores = ncores)
    
    res <- merge(data.frame(row.names = geneset, 
                            rna_log2FoldChange=de[,"log2FoldChange"],
                            rna_padj=de[,"padj"],
                            rna_abs_med_diff=de[,"diff"]),
                 do.call(rbind, res), by=0)
    
    rownames(res) <- res[,1]
    res <- res[,-1]
    
    toFormat          <- sapply(colnames(res), function(cn) !grepl("padj", cn) & is.numeric(res[,cn]))
    res[,toFormat]    <- round(res[,toFormat],digits=2)
    res[,"rna_padj"]  <- round(res[,"rna_padj"], digits=6)
    res[,"quadrant"]  <- getQuadrant(res$rna_log2FoldChange, res$atac_log2FoldChange)
    
    lfc[[name]] <- res
  }
  return(lfc)
}

```

# Compute correlation correlation coefficients for all genes and their TSS signal

```{r}

##################################################
## IF RSTUDIO HANGS RUN THIS AS SEPARATE SCRIPT ##
##################################################

target <- list(PCvBPH   = list(rna_PCvBPH,   contrast_PCvBPH),
               CRPCvPC  = list(rna_CRPCvPC,  contrast_CRPCvPC))

lfc_bg <- doCor(target)

par(mfrow=c(2,2))
hist(lfc_bg$PCvBPH$pearson, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg$PCvBPH$pearson, na.rm=TRUE)))
hist(lfc_bg$PCvBPH$spearman, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg$PCvBPH$spearman, na.rm=TRUE)))
hist(lfc_bg$CRPCvPC$pearson, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg$CRPCvPC$pearson, na.rm=TRUE)))
hist(lfc_bg$CRPCvPC$spearman, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg$CRPCvPC$spearman, na.rm=TRUE)))
```

## Do barplot

```{r}
pdf("pictures//scatterplot_correlation_turp/tss_deg_v3/per_sector_counts_bg.pdf")
par(mfrow=c(2,1), oma = c(0,0,1,0))
for (i in seq_along(lfc_bg)) {
  df <- lfc_bg[[i]]
  comparison <- names(lfc_bg)[i]
  doBarplot(df$quadrant, comparison)
}
mtext("Number of genes per sector of log2 fold change scatterplot", outer = T, line = 0)
dev.off()
par(op)
```

# Compute correlation correlation coefficients for DE genes and their TSS signal

```{r}

##################################################
## IF RSTUDIO HANGS RUN THIS AS SEPARATE SCRIPT ##
##################################################

contrast_PCvBPH <- c("PC", "BPH")
contrast_CRPCvPC <- c("CRPC", "PC")

target <- list(PCvBPH   = list(de_PCvBPH,   contrast_PCvBPH),
               CRPCvPC  = list(de_CRPCvPC,  contrast_CRPCvPC))

lfc <- doCor(target)

for (i in seq_along(lfc)) {
  
  cat(names(lfc)[i], "\n")
  
  sumatac <- sum(lfc[[i]][,"atac_log2FoldChange"] == 0)
  sumrna <-  sum(lfc[[i]][,"rna_log2FoldChange"] == 0)
  
  cat(sumatac, "/", nrow(lfc[[i]]), "ATAC log2 fold change are zero.\n")
  cat(sumrna, "/", nrow(lfc[[i]]), "RNA log2 fold change are zero.\n")
  
}

```

## Join with annotation table and dump results

```{r}
# dar.path <- "dar/turp_corrected_log2ratio_2/"
ann.gene <- as.data.frame(ann[ann$type == "gene" & ann$gene_id %in% rownames(rna_dds)])
annotated <- lapply(seq_along(lfc), function(i) {
  df <- lfc[[i]]
  df$gene_id <- rownames(df)
  
  
  # contrast <- names(lfc)[i]
  # fn <- tolower(gsub("([A-Z]+)v([A-Z]+)","dars_\\1_\\2.bed", contrast))
  
  merge(df,ann.gene,by="gene_id", all.x = TRUE)
})
names(annotated) <- names(lfc)
```

```{r}
library(openxlsx)
output.dir <- "tables/correlation_turp/tss_deg_v3"
if(!dir.exists(output.dir)) dir.create(output.dir, recursive = TRUE)
write.xlsx(x = annotated, file = file.path(output.dir, "degDar_correlation.xlsx"))
# for (i in seq_along(annotated)) {
  # write.csv(annotated[[i]], file = file.path(output.dir, paste0(names(annotated)[[i]],".csv")), row.names = FALSE)
# }
```

## Plot correlation coefficient distributions, count genes per sector and do barplot

```{r}
doHist <- function(vect, main) {
  m <- median(vect, na.rm = TRUE)
  hist(vect, breaks = "fd", main = main, xlab = sprintf("Correlation coefficient - median = %.2f",m), col = "grey90")
  abline(v=m, lwd = 2, col = "dodgerblue")
}

doBarplot <- function (vect, main) {
  t <- table(vect[!is.na(vect)])
  xlim <- c(0, 1.1*max(t))
  bplot <- barplot(t, main = main, horiz = TRUE, width = 0.85, xlim = xlim, yaxt="n")
  text(x = t, y = bplot, label = t, pos = 4, cex = 0.8)
  axis(2, at = bplot, labels = c("1"="Upper right (1)", "2"="Lower right (2)", "3"="Lower left (3)", "4"="Upper left (4)"), tick=FALSE, las=1, line=-0.5, cex.axis=0.65)
}


library(moments)

op <- par()
pdf("pictures//scatterplot_correlation_turp/tss_deg_v3/corcoeff_hists.pdf")
par(mfrow=c(2,2))
for (i in seq_along(lfc)) {
  df <- lfc[[i]]
  comparison <- names(lfc)[i]
  doHist(df$pearson, sprintf("%s - pearson", comparison))
  doHist(df$spearman, sprintf("%s - spearman", comparison))
  
  print(comparison)
  cat("Pearson range", range(df$pearson, na.rm = TRUE), "-", 
      "Skewness:", round(skewness(df$pearson, na.rm = TRUE),2),"-", 
      "Kurtosis:",round(kurtosis(df$pearson, na.rm = TRUE),2), "-", 
      "Excess kurtosis:",round(kurtosis(df$pearson, na.rm = TRUE)-3,2), "\n")
  cat("Spearman range:", range(df$spearman, na.rm = TRUE), "-", 
      "Skewness:", round(skewness(df$spearman, na.rm = TRUE),2), "-", 
      "Kurtosis:", round(kurtosis(df$spearman, na.rm = TRUE),2),"-",
      "Excess kurtosis:", round(kurtosis(df$spearman, na.rm = TRUE)-3,2),"\n")
  
}
dev.off()
par(op)
```

```{r}
pdf("pictures//scatterplot_correlation_turp/tss_deg_v3/per_sector_counts_v2.pdf")
par(mfrow=c(2,1), oma = c(0,0,2,0))

for(i in seq_along(lfc)) {
  quadrant_bg <- table(lfc_bg[[i]][["quadrant"]])
  quadrant_bg <- quadrands_bg / sum(quadrant_bg)
  
  quadrant_de <- table(lfc[[i]][["quadrant"]])
  quadrant_de <- quadrant_de / sum(quadrant_de)
  
  mat <- rbind(DEG=quadrant_de, all_genes=quadrant_bg)
  ylim <- c(0, 1.1 * max(mat))
  bp <- barplot(mat, beside = T, legend.text = gsub("_", " ", Hmisc::capitalize(rownames(mat))), xaxt = "n", ylab="Relative frequency", ylim = ylim, main = names(lfc)[i], col = c("red", "gray70"))
  axis(side = 1, las = 2, labels = c("Upper right", "Lower right", "Lower left", "Upper left"), at = apply(bp,2,mean), cex.axis = 0.7, tick = FALSE)
}
mtext("Relative abundance of genes for each sector of log2 fold change scatterplot", outer = T, line = 0)
dev.off()
par(op)
```

```{r}
pdf("pictures//scatterplot_correlation_turp/tss_deg_v3/quadrants_lfc_dens.pdf", height = 7, width = 10)
par(mfrow=c(1,2), oma = c(0,0,2,0))
for(i in seq_along(lfc)) {
  df <- lfc[[i]]
  df_all <- lfc_bg[[i]]
  
  dens1_deg <- density(df[df$quadrant == 1,"atac_log2FoldChange"])
  dens3_deg <- density(df[df$quadrant == 3,"atac_log2FoldChange"])
  densAll_deg <- density(df[,"atac_log2FoldChange"])
  
  dens1_all <- density(df_all[df_all$quadrant == 1,"atac_log2FoldChange"])
  dens3_all <- density(df_all[df_all$quadrant == 3,"atac_log2FoldChange"])
  densAll_all <- density(df_all[,"atac_log2FoldChange"])
  
  plot(densAll_all, col = "red", 
       ylim = c(0, max(dens1_deg$y, dens3_deg$y, densAll_deg$y, dens1_all$y, dens3_all$y, densAll_all$y)), 
       xlab = "ATAC-seq log2 fold change at TSS", main = names(lfc)[i],
       xlim  = quantile(c(dens1_deg$x, dens3_deg$x, densAll_deg$x, dens1_all$x, dens3_all$x, densAll_all$x), c(0.20, 0.9)))
  lines(dens3_all, col = "green")
  lines(dens1_all, col = "blue")
  
  lines(densAll_deg, col = "red", lty = 2)
  lines(dens1_deg, col = "blue", lty = 2)
  lines(dens3_deg, col = "green", lty = 2)
  
  legend("topright", 
         legend = c("All - all genes", "Upper right - all genes", "Lower left - all genes", "All - DEG", "Upper right - DEG", "Lower left - DEG"), 
         col = c("red", "blue", "green"), lwd = 1, title = "Quadrant", 
         cex = 0.6, lty = c(1,1,1,2,2,2))  
}
mtext("Comparison of log2 fold change across scatterplot quadrants", outer = T, line = 0)
dev.off()
```


# Do log2 fold change scatterplot

```{r}
source("R/doLFCscatterplot.R")
output.dir <- "pictures/scatterplot_correlation_turp/tss_deg_v3/"

ps <- list()
for (n in seq_along(lfc_bg)) {
  comparison <- names(lfc_bg)[n]
  
  df <- lfc_bg[[comparison]][,c("atac_log2FoldChange", "rna_log2FoldChange")]
  colnames(df) <- c("x", "y")
  df$type <- "bg"
  df[rownames(df) %in% rownames(lfc[[n]]), "type"] <- "de"
  
  
  alpha <- rep(0.1, nrow(df))
  alpha[df$type == "de"] <- 1
  ps[[n]] <- doLFCscatter(df1 = df,
                          ptitle1 = sprintf("%s (%d genes)", comparison, nrow(df)),
                          return_grob = TRUE, ythr = NULL, colors = c("grey70", "red"), 
                          palpha = alpha, legend_title = "", bilayer = TRUE)
}

gp          <- get.gpar()
gp$fontsize <- 14
gp$fontface <- "bold"
gp$cex      <- 1.1
ggsave(marrangeGrob(grobs = ps,
                    nrow = 1, ncol = length(lfc),
                    top = textGrob(label = "ATAC-seq log2 fold change at proximal promoter area (from -500bp to 500bp) vs RNA-seq log2 fold change",
                                   gp=gp)),
       filename=file.path(output.dir,"plot_bg_v2.pdf"), width=15, height=7)
```









