---
title: 'Tampere PC: Gene expression vs ATAC signal around TSS correlation (all genes and DEG)'
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Francesco Tabaro"
output:
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: true  
    df_print: tibble
    fig_caption: true
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
    fig_caption: true
    df_print: paged
    toc_float:
      collapsed: false
      smooth_scroll: false
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir="/bmt-data/genomics//projects//atac_workdir/")

NCORES <- floor(parallel::detectCores() * 0.8)
```

```{r, include=FALSE, echo=FALSE}
library(data.table)
library(DESeq2)
library(gridExtra)
library(matrixStats)
library(ggplot2)
```

# Load data

## Annotation

Load Ensembl v90 GTF annotation.

```{r}
source("R/getAnnotationTable.R")
ann <- getAnnotationTable()
ann.gene.df <- as.data.frame(subset(ann, type == "gene"))
```

## RNA

```{r}
load("R_data/RNAdds.RData")
load("R_data/RNA_differentiallyExpressed.RData")
```

## sRNA

```{r}
load("R_data/sRNA_dds.RData")
load("R_data/sRNA_differentiallyExpressed.RData")
rownames(srna_se_filter) <- rowRanges(srna_se_filter)$Name
colnames(DE$PCvBPH)  <- c("diff", "log2FoldChange", "padj")
colnames(DE$CRPCvPC) <- c("diff", "log2FoldChange", "padj")
```

## Proteins

```{r}
prot_rc <- fread("tables/swissprotdata_normalized.tsv")

biomart_map <- fread("tables/uniprot2ensembleGeneID/biomart_missing_uniprot.tsv", select = c(1,2)) # mapping from Biomart (Ensembl 90)
names(biomart_map) <- c("uniprot", "ensgid")
biomart_map[, ensgid := paste(ensgid, collapse = "|"), by = uniprot]

uniprot_map <- fread("tables/uniprot2ensembleGeneID/uniprot_mapping_missed.tsv") # mapping from Uniprot (unknown Ensembl)
names(uniprot_map) <- c("uniprot", "ensgid")
uniprot_map[, ensgid := paste(ensgid, collapse = "|"), by = uniprot]

missing_map <- rbind(biomart_map, uniprot_map)
missing_map <- missing_map[!duplicated(missing_map)]
missing_map[, ensgid := sapply(ensgid, function(x) ifelse(any(grepl(x, ann.gene.df$gene_id)), grep(x, ann.gene.df$gene_id, value = TRUE), ""))]
missing_map <- missing_map[ensgid != ""]

all_prot_map <- prot_rc[,.(uniprot, gn)]
all_prot_map[, ensgid := ann.gene.df$gene_id[match(gn, ann.gene.df$gene_name)]]
all_prot_map[match(missing_map$uniprot, uniprot), ensgid := missing_map[match(uniprot, uniprot), ensgid]]

prot_rc <- merge(all_prot_map, prot_rc, by = c("uniprot", "gn"), all.y = TRUE)

```


## ATAC

Load ATAC-seq signal at TSS of quantified genes.

```{r}
ATAC_MAT_PATH    <- "peak_tss_dar_quantifications/tss_quantification.tsv.gz"

atac_rc_mat     <- fread(cmd = sprintf("zcat %s", ATAC_MAT_PATH), sep ="\t", header = TRUE, showProgress = FALSE)
prom_gid        <- fread("tss_grid/all_genes_500u500d_sorted.bed")
names(prom_gid) <- c("chrom", "start", "end", "gid", "score", "strand")

atac_rc_mat     <- merge(prom_gid, atac_rc_mat, by=c("chrom", "start", "end"), all.x=TRUE)
atac_rc_mat[,`:=`(score=NULL,strand=NULL)]
setkey(atac_rc_mat, gid)

# atac_rc_mat[,`:=`(dar_id=paste(chrom, start, end, sep="_"))]
# setkey(atac_rc_mat, dar_id)
```


## Find common set of samples

```{r}

atac.samples <- list(bph  = grep("^BPH",   names(atac_rc_mat), value = TRUE),       # 11
                     pc   = grep("^PC",    names(atac_rc_mat), value = TRUE),       # 16
                     crpc = grep("^CRPC", names(atac_rc_mat), value = TRUE))        # 11

shared.samples.rna <- intersect(names(atac_rc_mat), colnames(rna_dds))
rna.samples <- list(bph  = grep("^BPH",  colnames(rna_dds), value = TRUE),          # 10
                    pc   = grep("^PC",   colnames(rna_dds), value = TRUE),          # 16
                    crpc = grep("^CRPC", colnames(rna_dds), value = TRUE))          # 11

shared.samples.mirna <- intersect(names(atac_rc_mat), colnames(srna_se_filter))     # 28
mirna.samples <- list(bph  = grep("BPH",  colnames(srna_se_filter), value = TRUE),  # 3
                      pc   = grep("^PC",  colnames(srna_se_filter), value = TRUE),  # 29
                      crpc = grep("CRPC", colnames(srna_se_filter), value = TRUE))  # 11

shared.samples.proteins <- intersect(names(atac_rc_mat), names(prot_rc))            # 26
prot.samples <- list(bph  = grep("BPH",  names(prot_rc), value = TRUE),             # 10
                     pc   = grep("^PC",  names(prot_rc), value = TRUE),             # 17
                     crpc = grep("CRPC", names(prot_rc), value = TRUE))             # 11

contrast_PCvBPH <- c("PC", "BPH")
contrast_CRPCvPC <- c("CRPC", "PC")

```

# Define helper functions to compute correlation coefficients

Correlation will be computed between normalized RNA-seq signal of genes and normalized ATAC-seq signal quantified in proximal promoter region defined from -500bp to +500bp around TSS.

```{r}
computeLFC <- function (rc, sample.set1, sample.set2) {
  log2(median(rc[sample.set1]) + 1) - log2(median(rc[sample.set2]) + 1)
}


computeCor <- function (gid, atac_rc, method, samples) {
  
  if(gid %in% rownames(rna_dds)){
    rna_rc <- assay(rna_dds,"norm.counts")[gid, match(samples,colnames(rna_dds))]
  } else if (any(grepl(sprintf("%s$",gid), rownames(srna_se_filter), ignore.case = TRUE))) {
    rna_rc <- assay(srna_se_filter, "norm.counts")[grep(sprintf("%s$",gid), rownames(srna_se_filter), ignore.case = TRUE), match(samples, colnames(srna_se_filter))]
  } else if (gid %in% prot_rc[["uniprot"]])   {
    rna_rc <- setNames(as.numeric(prot_rc[uniprot == gid, ..samples]), samples)
  } else {
    stop(sprintf("Could not find %s neither in RNA-seq matrix, sRNA-seq matrix and protein matrix.", gid))
  }
  
  atac_rc <- atac_rc[match(samples, names(atac_rc))]
  stopifnot(all(names(atac_rc) == names(rna_rc)))
  
  tryCatch({
    
    return(cor(atac_rc, rna_rc, method=method, use="na.or.complete"))
    
  }, error = function (err) {
    
    print(length(atac_rc))
    print(class(atac_rc))
    
    print(gid)
    print(dim(rna_rc))
    print(length(rna_rc))
    print(class(rna_rc))
    
    stop(err)
  })
  
}

getQuadrant <- function (rna_lfc, atac_lfc) {
  apply(cbind(atac_lfc, rna_lfc),1,function(row){
    quadrant <- NA
    if(any(is.na(row))) return(quadrant)
    if(row[1] >= 0 & row[2] >= 0) {
      quadrant <- 1
    } else if (row[1] >= 0 & row[2] < 0) {
      quadrant <- 2
    } else if (row[1] < 0 & row[2] < 0) {
      quadrant <- 3
    } else if (row[1] < 0 & row[2] >= 0) {
      quadrant <- 4
    }
    return(quadrant)
  })
}

doCor <- function (target, s, ncores = parallel::detectCores()) {
  
  lfc <- vector("list", length = length(target))
  names(lfc) <- names(target)
  
  for (name in names(target)) {
    
    x <- target[[name]]
    
    de <- x[[1]]
    geneset <- rownames(de)
    
    s1 <- atac.samples[[tolower(x[[2]][1])]]
    s2 <- atac.samples[[tolower(x[[2]][2])]]
    sele_cols <- c(s1,s2)
    
    # s <- c(shared.samples[[tolower(x[[2]][1])]],
    #        shared.samples[[tolower(x[[2]][2])]])
    # s <- Reduce(c, shared.samples) # always use all samples to compute correlation
    
    res <- mclapply(seq_along(geneset), function (i) {

      gene    <- geneset[i]
      
      if (gene %in% prot_rc[["uniprot"]]) {
        gene <- prot_rc[uniprot == gene, "ensgid"]
      }
      
      if (!is.na(gene)){
        tss_rc  <- setNames(as.numeric(atac_rc_mat[gid == gene, ..s]), s)
        pearson  <- computeCor(geneset[i], tss_rc, "pearson", s)
        spearman <- computeCor(geneset[i], tss_rc, "spearman", s)
        
        tss_rc  <- setNames(as.numeric(atac_rc_mat[gid == gene, ..sele_cols]), sele_cols)
        lfc     <- computeLFC(tss_rc, s1, s2)
        
      } else { # some proteins do not map on ensembl
        pearson  <- NA
        spearman <- NA
        lfc      <- NA
        
      }
      
      return(data.frame(row.names=geneset[i], # gene variable change sometimes...
                        atac_log2FoldChange=lfc,
                        pearson=pearson,
                        spearman=spearman))
    }, mc.cores = ncores)

    res <- merge(data.frame(row.names = geneset, 
                            rna_log2FoldChange=de[,"log2FoldChange"],
                            rna_padj=de[,"padj"],
                            rna_abs_med_diff=de[,"diff"]),
                 do.call(rbind, res), by=0)
    
    rownames(res) <- res[,1]
    res <- res[,-1]
    
    toFormat          <- sapply(colnames(res), function(cn) !grepl("padj", cn) & is.numeric(res[,cn]))
    res[,toFormat]    <- round(res[,toFormat],digits=2)
    res[,"rna_padj"]  <- round(res[,"rna_padj"], digits=6)
    res[,"quadrant"]  <- getQuadrant(res$rna_log2FoldChange, res$atac_log2FoldChange)
    
    lfc[[name]] <- res
  }
  return(lfc)
}

```

# Compute correlation correlation coefficients for all genes and their TSS signal

## RNA-se

```{r}

##################################################
## IF RSTUDIO HANGS RUN THIS AS SEPARATE SCRIPT ##
##################################################

target <- list(PCvBPH   = list(rna_PCvBPH,   contrast_PCvBPH),
               CRPCvPC  = list(rna_CRPCvPC,  contrast_CRPCvPC))

lfc_bg <- doCor(target, shared.samples.rna, NCORES)

par(mfrow=c(2,2))
hist(lfc_bg$PCvBPH$pearson, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg$PCvBPH$pearson, na.rm=TRUE)))
hist(lfc_bg$PCvBPH$spearman, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg$PCvBPH$spearman, na.rm=TRUE)))
hist(lfc_bg$CRPCvPC$pearson, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg$CRPCvPC$pearson, na.rm=TRUE)))
hist(lfc_bg$CRPCvPC$spearman, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg$CRPCvPC$spearman, na.rm=TRUE)))
```

## sRNA-seq

```{r}

doMirnaDf <- function (i, s1, s2) {
  row <- assay(srna_se_filter, "norm.counts")[i, ]
  data.frame(row.names = rownames(srna_se_filter)[i],
    log2FoldChange = log2(median(row[s1])) - log2(median(row[s2])),
    diff = median(row[s1]) - median(row[s2]),
    padj = wilcox.test(row[s1], row[s2], exact = FALSE)$p.value)
}

bg_mirna_PCvBPH  <- do.call(rbind, mclapply(seq(nrow(srna_se_filter)), doMirnaDf, s1 = mirna.samples$pc, s2 = mirna.samples$bph, mc.cores = NCORES))
bg_mirna_PCvBPH$padj <- p.adjust(bg_mirna_PCvBPH$padj, "fdr")

bg_mirna_CRPCvPC <- do.call(rbind, mclapply(seq(nrow(srna_se_filter)), doMirnaDf, s1 = mirna.samples$crpc, s2 = mirna.samples$pc, mc.cores = NCORES))
bg_mirna_CRPCvPC$padj <- p.adjust(bg_mirna_CRPCvPC$padj, "fdr")

```

```{r}

target <- list(PCvBPH  = list(DataFrame(bg_mirna_PCvBPH),  contrast_PCvBPH),
               CRPCvPC = list(DataFrame(bg_mirna_CRPCvPC), contrast_CRPCvPC))

lfc_bg_mirna <- doCor(target, shared.samples.mirna, NCORES)

par(mfrow=c(2,2))
hist(lfc_bg_mirna$PCvBPH$pearson, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg_mirna$PCvBPH$pearson, na.rm=TRUE)))
hist(lfc_bg_mirna$PCvBPH$spearman, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg_mirna$PCvBPH$spearman, na.rm=TRUE)))
hist(lfc_bg_mirna$CRPCvPC$pearson, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg_mirna$CRPCvPC$pearson, na.rm=TRUE)))
hist(lfc_bg_mirna$CRPCvPC$spearman, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_bg_mirna$CRPCvPC$spearman, na.rm=TRUE)))
```
## RNA-seq and sRNA-seq

```{r}
pear1 <- c(lfc_bg_mirna$PCvBPH$pearson, lfc_bg$PCvBPH$pearson)
pear2 <- c(lfc_bg_mirna$CRPCvPC$pearson, lfc_bg$CRPCvPC$pearson)
  
spear1 <- c(lfc_bg_mirna$PCvBPH$spearman, lfc_bg$PCvBPH$spearman)
spear2 <- c(lfc_bg_mirna$CRPCvPC$spearman, lfc_bg$CRPCvPC$spearman)
  
par(mfrow=c(2,2))
hist(pear1, breaks = "fd", xlab = sprintf("Pearson correlation coeff; med = %.3f; sd = %.3f", median(pear1, na.rm=TRUE), sd(pear1, na.rm = T)))
hist(spear1, breaks = "fd", xlab = sprintf("Spearman correlation coeff; med = %.3f; sd = %.3f", median(spear1, na.rm=TRUE), sd(spear1, na.rm = T)))
hist(pear2, breaks = "fd", xlab = sprintf("Pearson coeff; med = %.3f; sd = %.3f", median(pear2, na.rm=TRUE), sd(pear2, na.rm = T)))
hist(spear2, breaks = "fd", xlab = sprintf("Spearman coeff; med = %.3f; sd = %.3f", median(spear2, na.rm=TRUE), sd(spear2, na.rm = T)))


```


## Proteins

```{r}

doProtDf <- function (i, s1, s2){
  k <- match(c(s1,s2), names(prot_rc))
  row <- setNames(as.numeric(prot_rc[i, ..k]), c(s1,s2))
  data.frame(row.names = prot_rc[["uniprot"]][i],
             log2FoldChange = log2(median(row[s1])) - log2(median(row[s2])),
             diff = median(row[s1]) - median(row[s2]),
             padj = wilcox.test(row[s1], row[s2], exact = FALSE)$p.value)
}

bg_prot_PCvBPH  <- do.call(rbind, mclapply(seq(nrow(prot_rc)), doProtDf, s1 = prot.samples$pc,   s2 = prot.samples$bph, mc.cores = NCORES))
bg_prot_CRPCvPC <- do.call(rbind, mclapply(seq(nrow(prot_rc)), doProtDf, s1 = prot.samples$crpc, s2 = prot.samples$pc,  mc.cores = NCORES))

```

```{r}
target <- list(PCvBPH  = list(DataFrame(bg_prot_PCvBPH), contrast_PCvBPH),
               CRPCvPC = list(DataFrame(bg_prot_CRPCvPC), contrast_CRPCvPC))

lfc_prot_bg <- doCor(target, shared.samples.proteins, NCORES)

par(mfrow=c(2,2))
hist(lfc_prot_bg$PCvBPH$pearson, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_prot_bg$PCvBPH$pearson, na.rm=TRUE)))
hist(lfc_prot_bg$PCvBPH$spearman, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_prot_bg$PCvBPH$spearman, na.rm=TRUE)))
hist(lfc_prot_bg$CRPCvPC$pearson, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_prot_bg$CRPCvPC$pearson, na.rm=TRUE)))
hist(lfc_prot_bg$CRPCvPC$spearman, breaks = "fd", xlab = sprintf("Pearson correlation coeff; median = %.3f", median(lfc_prot_bg$CRPCvPC$spearman, na.rm=TRUE)))

print(sd(lfc_prot_bg$PCvBPH$pearson, na.rm=TRUE))
print(sd(lfc_prot_bg$PCvBPH$spearman, na.rm=TRUE))
```


## Correlation of proteins whose coding gene correlate with ATAC-seq

### Add mapping to Ensembl gene id to protein log fold change tables

```{r}
map <- as.data.frame(prot_rc[,.(uniprot, ensgid)])
for (i in seq_along(lfc_prot_bg)) {
  df <- lfc_prot_bg[[i]]
  df <- merge(df, map, by.x = 0, by.y = "uniprot")
  rownames(df) <- df$Row.names
  df <- df[,-1]
  
  lfc_prot_bg[[i]] <- df
}
```


### Find genes correlating over threshold

#### Test different thresholds

```{r}
mrna <- lfc_bg[[1]]
prot <- lfc_prot_bg[[1]]

```


```{r}
COR_THR <- seq(0.2, 1, by=0.1)

mats <- list(mrna_only=matrix(0, length(COR_THR), 2),
             prot_only=matrix(0, length(COR_THR), 2),
             intersection=matrix(0, length(COR_THR), 2))

for ( i in seq_along(COR_THR)){
  thr <- COR_THR[i]
  
  smrna <- subset(mrna, abs(pearson) > thr | abs(spearman) > thr)
  sprot <- subset(prot, abs(pearson) > thr | abs(spearman) > thr)
  
  if(nrow(smrna) > 0 && nrow(sprot) > 0) {

    part <- get.venn.partitions(list(mrna=rownames(smrna),proteins=sprot$ensgid))
    
    intersection <- subset(part, ..set.. == "mrna∩proteins", select = ..values..)
    intersection <- intersection$..values..$`1`
    mats$intersection[i,] <- c(thr, length(intersection))
    
    mrna_only <- subset(part, ..set.. == "(mrna)∖(proteins)", select = ..values..)
    mrna_only <- mrna_only$..values..$`3`
    mats$mrna_only[i,] <- c(thr, length(mrna_only))  
    
    prot_only <- subset(part, ..set.. == "(proteins)∖(mrna)", select = ..values..)
    prot_only <- prot_only$..values..$`2`
    mats$prot_only[i,] <- c(thr, length(prot_only))
    
  } else if ( nrow(smrna) == 0 & nrow(sprot) > 0) {
    
    mats$mrna_only[i,] <- c(thr, 0)
    mats$prot_only[i,] <- c(thr, nrow(sprot))
    mats$intersection[i,] <- c(thr, 0)
    
  } else if ( nrow(smrna) > 0 & nrow(sprot) == 0) {
    
    mats$mrna_only[i,] <- c(thr, nrow(smrna))
    mats$prot_only[i,] <- c(thr, 0)
    mats$intersection[i,] <- c(thr, 0)
    
  } else if ( nrow(smrna) == 0 & nrow(sprot) == 0) {
    
    mats$mrna_only[i,] <- c(thr, 0)
    mats$prot_only[i,] <- c(thr, 0)
    mats$intersection[i,] <- c(thr, 0)
    
  }else {
    stop("WTF")
  }
}

mats <- lapply(mats, function(each) { each[,2] <- each[,2] + 1; return(each) })

ylim <- c(0, max(sapply(mats,function(each) each[,2])))
plot(mats[[1]], col = "red", lty = 1, type  = "l", log = "y", 
     xlab = "Correlation coefficient threshold (Pearson or Spearman)", ylab = "log2 number of genes",
     xaxt = "n", las = 2)
lines(mats[[2]], col = "green")
lines(mats[[3]], col = "blue")
axis(1, at=COR_THR, labels=COR_THR)
legend("toprigh", col = c("red", "green", "blue"), legend = c("mrna only", "protein only", "intersection"), lwd = 1)

ylim <- c(0, max(sapply(mats,function(each) each[,2])))
plot(mats[[1]], col = "red", lty = 1, type  = "l", 
     xlab = "Correlation coefficient threshold (Pearson or Spearman)", ylab = "number of genes",
     xaxt = "n", las = 2)
lines(mats[[2]], col = "green")
lines(mats[[3]], col = "blue")
axis(1, at=COR_THR, labels=COR_THR)
legend("toprigh", col = c("red", "green", "blue"), legend = c("mrna only", "protein only", "intersection"), lwd = 1)



```


```{r}
library(VennDiagram)
library(openxlsx)
COR_THR <- 0.5

# par(mfrow=c(2,2))
# for (i in seq_along(lfc_bg)) {
#   # print(names(lfc_bg)[i])
#   # 
#   # df <- lfc_bg[[i]]
#   # print(dim(df))
#   # sdf <- subset(df, abs(pearson) > COR_THR | abs(spearman) > COR_THR)
#   # print(dim(sdf))
#   # 
#   # genes <- rownames(sdf)
#   # 
#   # pdf <- lfc_prot_bg[[i]]
#   # print(dim(pdf))
#   # pdf <- pdf[pdf$ensgid %in% genes,]
#   # print(dim(pdf))
#   # 
#   # hist(pdf$pearson, breaks = "fd", xlab = sprintf("Pearson coeff; median = %.3f; sd = %.3f", median(pdf$pearson, na.rm=TRUE), sd(pdf$pearson, na.rm=TRUE)))
#   # hist(pdf$spearman, breaks = "fd", xlab = sprintf("Spearman coeff; median = %.3f; sd = %.3f", median(pdf$spearman, na.rm=TRUE), sd(pdf$spearman, na.rm=TRUE)))
# 
# }

smrna <- subset(mrna, abs(pearson) > COR_THR | abs(spearman) > COR_THR)
sprot <- subset(prot, abs(pearson) > COR_THR | abs(spearman) > COR_THR)

venn.diagram(list(mrna=rownames(smrna),proteins=sprot$ensgid), filename = "~/test/vd.tiff")

part <- get.venn.partitions(list(mrna=rownames(smrna),proteins=sprot$ensgid))

intersection <- subset(part, ..set.. == "mrna∩proteins", select = ..values..)
intersection <- intersection$..values..$`1`
intersection <- subset(ann, subset = type == "gene" & gene_id %in% intersection, select = gene_name)$gene_name

mrna_only <- subset(part, ..set.. == "(mrna)∖(proteins)", select = ..values..)
mrna_only <- mrna_only$..values..$`3`
mrna_only <- subset(ann, subset = type == "gene" & gene_id %in% mrna_only, select = gene_name)$gene_name

prot_only <- subset(part, ..set.. == "(proteins)∖(mrna)", select = ..values..)
prot_only <- prot_only$..values..$`2`
prot_only <- subset(ann, subset = type == "gene" & gene_id %in% prot_only, select = gene_name)$gene_name

write.xlsx(x = list(mrna_only=as.data.frame(mrna_only), 
                    protein_only=as.data.frame(prot_only), 
                    intersection=as.data.frame(intersection)),
           file = "/bmt-data/genomics/projects/atac_workdir/tables/gene_and_proteins_correlating_with_atac_at_TSS.xls")

```

# Compute correlation correlation coefficients for DE genes and their TSS signal

```{r}

##################################################
## IF RSTUDIO HANGS RUN THIS AS SEPARATE SCRIPT ##
##################################################

target <- list(PCvBPH   = list(de_PCvBPH,   contrast_PCvBPH),
               CRPCvPC  = list(de_CRPCvPC,  contrast_CRPCvPC))

lfc <- doCor(target, shared.samples.rna, NCORES)

for (i in seq_along(lfc)) {
  
  cat(names(lfc)[i], "\n")
  
  sumatac <- sum(lfc[[i]][,"atac_log2FoldChange"] == 0)
  sumrna <-  sum(lfc[[i]][,"rna_log2FoldChange"] == 0)
  
  cat(sumatac, "/", nrow(lfc[[i]]), "ATAC log2 fold change are zero.\n")
  cat(sumrna, "/", nrow(lfc[[i]]), "RNA log2 fold change are zero.\n")
  
}

```

## Join with annotation table and dump results

```{r}
# dar.path <- "dar/turp_corrected_log2ratio_2/"
ann.gene <- as.data.frame(ann[ann$type == "gene" & ann$gene_id %in% rownames(rna_dds)])
annotated <- lapply(seq_along(lfc), function(i) {
  df <- lfc[[i]]
  df$gene_id <- rownames(df)
  
  
  # contrast <- names(lfc)[i]
  # fn <- tolower(gsub("([A-Z]+)v([A-Z]+)","dars_\\1_\\2.bed", contrast))
  
  merge(df,ann.gene,by="gene_id", all.x = TRUE)
})
names(annotated) <- names(lfc)
```

### Save tables

```{r}
library(openxlsx)
output.dir <- "tables/correlation_turp/tss_deg_v3"
if(!dir.exists(output.dir)) dir.create(output.dir, recursive = TRUE)
write.xlsx(x = annotated, file = file.path(output.dir, "degTss_correlation.xlsx"))
# for (i in seq_along(annotated)) {
  # write.csv(annotated[[i]], file = file.path(output.dir, paste0(names(annotated)[[i]],".csv")), row.names = FALSE)
# }
```

```{r}
annotated <- lapply(seq_along(lfc), function(i) {
  df <- lfc_bg[[i]]
  df$gene_id <- rownames(df)
  
  
  # contrast <- names(lfc)[i]
  # fn <- tolower(gsub("([A-Z]+)v([A-Z]+)","dars_\\1_\\2.bed", contrast))
  
  merge(df,ann.gene,by="gene_id", all.x = TRUE)
})
names(annotated) <- names(lfc)
write.xlsx(annotated, file = file.path(output.dir, "allTss_correlation.xlsx"))
```

```{r}
annotated <- lapply(seq_along(lfc), function(i) {
  df <- lfc_prot_bg[[i]]
  df$gene_id <- rownames(df)
  
  
  # contrast <- names(lfc)[i]
  # fn <- tolower(gsub("([A-Z]+)v([A-Z]+)","dars_\\1_\\2.bed", contrast))
  
  merge(df,ann.gene,by.x = "ensgid", by.y = "gene_id", all.x = TRUE)
})
names(annotated) <- names(lfc)
write.xlsx(annotated, file = file.path(output.dir, "allProtTss_correlation.xlsx"))
```


# Plot correlation coefficient distributions, count genes per sector and do barplot

```{r}
doHist <- function(vect, main) {
  m <- median(vect, na.rm = TRUE)
  hist(vect, breaks = "fd", main = main, xlab = sprintf("Correlation coefficient - median = %.2f",m), col = "grey90")
  abline(v=m, lwd = 2, col = "dodgerblue")
}

doBarplot <- function (vect, main) {
  t <- table(vect[!is.na(vect)])
  xlim <- c(0, 1.1*max(t))
  bplot <- barplot(t, main = main, horiz = TRUE, width = 0.85, xlim = xlim, yaxt="n")
  text(x = t, y = bplot, label = t, pos = 4, cex = 0.8)
  axis(2, at = bplot, labels = c("1"="Upper right (1)", "2"="Lower right (2)", "3"="Lower left (3)", "4"="Upper left (4)"), tick=FALSE, las=1, line=-0.5, cex.axis=0.65)
}


library(moments)

op <- par()
pdf("pictures//scatterplot_correlation_turp/tss_deg_v3/corcoeff_hists.pdf")
par(mfrow=c(2,2))
for (i in seq_along(lfc)) {
  df <- lfc[[i]]
  comparison <- names(lfc)[i]
  doHist(df$pearson, sprintf("%s - pearson", comparison))
  doHist(df$spearman, sprintf("%s - spearman", comparison))
  
  print(comparison)
  cat("Pearson range", range(df$pearson, na.rm = TRUE), "-", 
      "Skewness:", round(skewness(df$pearson, na.rm = TRUE),2),"-", 
      "Kurtosis:",round(kurtosis(df$pearson, na.rm = TRUE),2), "-", 
      "Excess kurtosis:",round(kurtosis(df$pearson, na.rm = TRUE)-3,2), "\n")
  cat("Spearman range:", range(df$spearman, na.rm = TRUE), "-", 
      "Skewness:", round(skewness(df$spearman, na.rm = TRUE),2), "-", 
      "Kurtosis:", round(kurtosis(df$spearman, na.rm = TRUE),2),"-",
      "Excess kurtosis:", round(kurtosis(df$spearman, na.rm = TRUE)-3,2),"\n")
  
}
dev.off()
par(op)
```

```{r}
pdf("pictures//scatterplot_correlation_turp/tss_deg_v3/per_sector_counts_v2.pdf")
par(mfrow=c(2,1), oma = c(0,0,2,0))

for(i in seq_along(lfc)) {
  quadrant_bg <- table(lfc_bg[[i]][["quadrant"]])
  quadrant_bg <- quadrands_bg / sum(quadrant_bg)
  
  quadrant_de <- table(lfc[[i]][["quadrant"]])
  quadrant_de <- quadrant_de / sum(quadrant_de)
  
  mat <- rbind(DEG=quadrant_de, all_genes=quadrant_bg)
  ylim <- c(0, 1.1 * max(mat))
  bp <- barplot(mat, beside = T, legend.text = gsub("_", " ", Hmisc::capitalize(rownames(mat))), xaxt = "n", ylab="Relative frequency", ylim = ylim, main = names(lfc)[i], col = c("red", "gray70"))
  axis(side = 1, las = 2, labels = c("Upper right", "Lower right", "Lower left", "Upper left"), at = apply(bp,2,mean), cex.axis = 0.7, tick = FALSE)
}
mtext("Relative abundance of genes for each sector of log2 fold change scatterplot", outer = T, line = 0)
dev.off()
par(op)
```

```{r}
pdf("pictures//scatterplot_correlation_turp/tss_deg_v3/quadrants_lfc_dens.pdf", height = 7, width = 10)
par(mfrow=c(1,2), oma = c(0,0,2,0))
for(i in seq_along(lfc)) {
  df <- lfc[[i]]
  df_all <- lfc_bg[[i]]
  
  dens1_deg <- density(df[df$quadrant == 1,"atac_log2FoldChange"])
  dens3_deg <- density(df[df$quadrant == 3,"atac_log2FoldChange"])
  densAll_deg <- density(df[,"atac_log2FoldChange"])
  
  dens1_all <- density(df_all[df_all$quadrant == 1,"atac_log2FoldChange"])
  dens3_all <- density(df_all[df_all$quadrant == 3,"atac_log2FoldChange"])
  densAll_all <- density(df_all[,"atac_log2FoldChange"])
  
  plot(densAll_all, col = "red", 
       ylim = c(0, max(dens1_deg$y, dens3_deg$y, densAll_deg$y, dens1_all$y, dens3_all$y, densAll_all$y)), 
       xlab = "ATAC-seq log2 fold change at TSS", main = names(lfc)[i],
       xlim  = quantile(c(dens1_deg$x, dens3_deg$x, densAll_deg$x, dens1_all$x, dens3_all$x, densAll_all$x), c(0.20, 0.9)))
  lines(dens3_all, col = "green")
  lines(dens1_all, col = "blue")
  
  lines(densAll_deg, col = "red", lty = 2)
  lines(dens1_deg, col = "blue", lty = 2)
  lines(dens3_deg, col = "green", lty = 2)
  
  legend("topright", 
         legend = c("All - all genes", "Upper right - all genes", "Lower left - all genes", "All - DEG", "Upper right - DEG", "Lower left - DEG"), 
         col = c("red", "blue", "green"), lwd = 1, title = "Quadrant", 
         cex = 0.6, lty = c(1,1,1,2,2,2))  
}
mtext("Comparison of log2 fold change across scatterplot quadrants", outer = T, line = 0)
dev.off()
```


# Do log2 fold change scatterplot

```{r}
source("R/doLFCscatterplot.R")
output.dir <- "pictures/scatterplot_correlation_turp/tss_deg_v3/"

ps <- list()
for (n in seq_along(lfc_bg)) {
  comparison <- names(lfc_bg)[n]
  
  df <- lfc_bg[[comparison]][,c("atac_log2FoldChange", "rna_log2FoldChange")]
  colnames(df) <- c("x", "y")
  df$type <- "bg"
  df[rownames(df) %in% rownames(lfc[[n]]), "type"] <- "de"
  
  
  alpha <- rep(0.1, nrow(df))
  alpha[df$type == "de"] <- 1
  ps[[n]] <- doLFCscatter(df1 = df,
                          ptitle1 = sprintf("%s (%d genes)", comparison, nrow(df)),
                          return_grob = TRUE, ythr = NULL, colors = c("grey70", "red"), 
                          palpha = alpha, legend_title = "", bilayer = TRUE)
}

gp          <- get.gpar()
gp$fontsize <- 14
gp$fontface <- "bold"
gp$cex      <- 1.1
ggsave(marrangeGrob(grobs = ps,
                    nrow = 1, ncol = length(lfc),
                    top = textGrob(label = "ATAC-seq log2 fold change at proximal promoter area (from -500bp to 500bp) vs RNA-seq log2 fold change",
                                   gp=gp)),
       filename=file.path(output.dir,"plot_bg_v2.pdf"), width=15, height=7)
```

## Poster version

```{r}
all_pc_bph <- readWorkbook(xlsxFile = "tables/correlation_turp/tss_deg_v3/allTss_correlation.xlsx", sheet = 1)
all_crpc_pc <- readWorkbook(xlsxFile = "tables/correlation_turp/tss_deg_v3/allTss_correlation.xlsx", sheet = 2)

de_pc_bph <- readWorkbook(xlsxFile = "tables/correlation_turp/tss_deg_v3/degDar_correlation.xlsx", sheet = 1)
de_crpc_pc <- readWorkbook(xlsxFile = "tables/correlation_turp/tss_deg_v3/degDar_correlation.xlsx", sheet = 2)

```

```{r}
dir.create("pictures/scatterplot_correlation_turp/tss_deg_v4")
pdf("pictures/scatterplot_correlation_turp/tss_deg_v4/plot.pdf", width = 14)

cex.annotations <- 1.4

par(mar = c(5,4,4,6) + 0.1, mfrow = c(1,2), oma = c(0,0,2,0) + 0.5, pty = "s", 
    cex.axis = cex.annotations, cex.main = cex.annotations,cex.lab = cex.annotations)

xlim <- c(-2,2)
ylim <- c(-10, 10)

x <- all_pc_bph$atac_log2FoldChange
y <- all_pc_bph$rna_log2FoldChange

x[x < xlim[1]] <- xlim[1]
x[x > xlim[2]] <- xlim[2]
y[y < ylim[1]] <- ylim[1]
y[y > ylim[2]] <- ylim[2]

plot(x,y, type = "n",
     xlab = "ATAC-seq log2 fold change",
     ylab = "RNA-seq log2 folde change",
     axes = FALSE)

xaxis <- pretty(xlim, n = 4)
yaxis <- pretty(ylim, n = 6)
rugx <- seq(xlim[1], xlim[2] - 0.5, by = 0.5)
rugy <- seq(ylim[1], ylim[2] - 2.5, by = 2.5)

xlabs <- rugx
xlabs[!xlabs %in% xaxis] <- NA
xlabs <- c(xlabs, xlim[2])
ylabs <- rugy
ylabs[!ylabs %in% yaxis] <- NA
ylabs <- c(ylabs, ylim[2])

Axis(side = 1, at = xaxis, labels = as.character(xaxis))
rug(x = rugx, ticksize = -0.01, side = 1, lwd = 1)
# rug(x = xaxis, ticksize = -0.03, side = 1, lwd = 1)

Axis(side = 2, at = yaxis, labels = as.character(yaxis), las = 2)
rug(x = rugy, ticksize = -0.01, side = 2, lwd = 1)
# rug(x = yaxis, ticksize = -0.03, side = 2, lwd = 1)

grid()
box()

points(x,y,cex = 0.2, pch = 19, col = "grey60")

x <- de_pc_bph$atac_log2FoldChange
y <- de_pc_bph$rna_log2FoldChange

x[x < xlim[1]] <- xlim[1]
x[x > xlim[2]] <- xlim[2]
y[y < ylim[1]] <- ylim[1]
y[y > ylim[2]] <- ylim[2]


points(x,y,col = "red", 
       pch = 19, cex = 0.2)

title("PC vs BPH")

abline(v=0,h=0)

text(1.5,9.5,sprintf("%d/%d", 
                     sum(de_pc_bph$atac_log2FoldChange > 0 & de_pc_bph$rna_log2FoldChange > 0),
                     sum(all_pc_bph$quadrant == 1)),
     cex = cex.annotations)

text(1.5,-9.5,sprintf("%d/%d", 
                     sum(de_pc_bph$atac_log2FoldChange > 0 & de_pc_bph$rna_log2FoldChange < 0),
                     sum(all_pc_bph$quadrant == 2)),
     cex = cex.annotations)

text(-1.5,-9.5,sprintf("%d/%d", 
                     sum(de_pc_bph$atac_log2FoldChange < 0 & de_pc_bph$rna_log2FoldChange < 0),
                     sum(all_pc_bph$quadrant == 3)),
     cex = cex.annotations)

text(-1.5,9.5,sprintf("%d/%d", 
                     sum(de_pc_bph$atac_log2FoldChange < 0 & de_pc_bph$rna_log2FoldChange > 0),
                     sum(all_pc_bph$quadrant == 4)),
     cex = cex.annotations)

legend(2.25, 10, 
       legend = c("All", "Deg"),
       col = c("grey40", "red"),
       pch = 19,
       xpd = TRUE,
       cex = cex.annotations)

x <- all_crpc_pc$atac_log2FoldChange
y <- all_crpc_pc$rna_log2FoldChange

x[x < xlim[1]] <- xlim[1]
x[x > xlim[2]] <- xlim[2]
y[y < ylim[1]] <- ylim[1]
y[y > ylim[2]] <- ylim[2]


plot(x,y, type = "n",
     xlab = "ATAC-seq log2 fold change",
     ylab = "RNA-seq log2 folde change",
     axes = FALSE)

Axis(side = 1, at = xaxis, labels = as.character(xaxis))
rug(x = rugx, ticksize = -0.01, side = 1, lwd = 1)
# rug(x = xaxis, ticksize = -0.03, side = 1, lwd = 1)

Axis(side = 2, at = yaxis, labels = as.character(yaxis), las = 2)
rug(x = rugy, ticksize = -0.01, side = 2, lwd = 1)
# rug(x = yaxis, ticksize = -0.03, side = 2, lwd = 1)

grid()
box()

points(x,y,cex = 0.2,
     pch = 19,
     xlab = "ATAC-seq log2 fold change",
     ylab = "RNA-seq log2 folde change",
     col = "grey60"
     # cex = cex
     )

x <- de_crpc_pc$atac_log2FoldChange
y <- de_crpc_pc$rna_log2FoldChange

x[x < xlim[1]] <- xlim[1]
x[x > xlim[2]] <- xlim[2]
y[y < ylim[1]] <- ylim[1]
y[y > ylim[2]] <- ylim[2]

points(x,y,col = "red", 
       pch = 19, cex = 0.2)

title("CRPC vs PC")
abline(v=0,h=0)

text(1.5,9.5,sprintf("%d/%d", 
                     sum(de_crpc_pc$atac_log2FoldChange > 0 & de_crpc_pc$rna_log2FoldChange > 0),
                     sum(all_crpc_pc$quadrant == 1)),
     cex = cex.annotations)

text(1.5,-9.5,sprintf("%d/%d", 
                     sum(de_crpc_pc$atac_log2FoldChange > 0 & de_crpc_pc$rna_log2FoldChange < 0),
                     sum(all_crpc_pc$quadrant == 2)),
     cex = cex.annotations)

text(-1.5,-9.5,sprintf("%d/%d", 
                     sum(de_crpc_pc$atac_log2FoldChange < 0 & de_crpc_pc$rna_log2FoldChange < 0),
                     sum(all_crpc_pc$quadrant == 3)),
     cex = cex.annotations)

text(-1.5,9.5,sprintf("%d/%d", 
                     sum(de_crpc_pc$atac_log2FoldChange < 0 & de_crpc_pc$rna_log2FoldChange > 0),
                     sum(all_crpc_pc$quadrant == 4)),
     cex = cex.annotations)

legend(2.25, 10, 
       legend = c("All", "Deg"),
       col = c("grey40", "red"),
       pch = 19,
       xpd = TRUE, 
       cex = cex.annotations)

title("Correlation of log2 fold changes at TSS", outer = TRUE, line = 1, cex = cex.annotations)
dev.off()
```

## Paper version

```{r}
library(ggrepel)
library(cowplot)
library(grid)
```


```{r}
# prostate cancer
interesting_genes <- c("KLK3",# aka PSA
                       "CDH1", "FOXA1", "CDKN1B", "ATM", "NKX3-1", "MED12", "PTEN", "RB1","AR")

# oncogenes
interesting_genes <- c(interesting_genes, "ABL1", "ABL2", "AKT1", "ATF1", "BCL11A", "BCL2", "BCL3", "BCL6", "BCR", "BRAF", "CARD11",
                       "CBLB", "CCND1", "CCND2", "CCND3", "CDX2", "CTNNB1", "DDB2", "DDIT3", "DDX6", "DEK", "EGFR",
                       "ELK4", "ERBB2", "ETV4", "ETV6", 
                       "MECOM",  #"EVI1", 
                       "EWSR1", "FEV", "FGFR1", "FGFR1OP", "FGFR2", "FUS", "GOLGA5", "GOPC", "HMGA1", "HMGA2", "HRAS",
                       "IRF4", "JUN", "KIT", "KRAS", "LCK", "LMO2", "MAF", "MAFB", "MAML2", "MDM2", "MET", "MITF", 
                       "KMT2A", # "MLL", 
                       "MPL", "MYB", "MYC", 
                       "MYCL", # "MYCL1", 
                       "MYCN", "NCOA4","NFKB2", "NRAS", "NTRK1", "NUP214", "PAX8", "PDGFB", "PIK3CA", "PIM1", "PLAG1",
                       "PPARG", "PTPN11","RAF1", "REL", "RET", "ROS1", "SMO", "SS18", "TCL1A", "TET2", "TFG", "TLX1",
                       "TPR", "USP6")
# tumor suppressor
interesting_genes <- c(interesting_genes, "APC", "ARHGEF12", "ATM", "BCL11B", "BLM", "BMPR1A", "BRCA1", "BRCA2", "CARS",
                       "CBFA2T3", "CDH1", "CDH11", "CDK6", "CDKN2C", "CEBPA", "CHEK2", "CREB1", "CREBBP", "CYLD", "DDX5",
                       "EXT1", "EXT2", "FBXW7", "FH", "FLT3", "FOXP1", "GPC3", "IDH1", "IL2", "JAK2", "MAP2K4", "MDM4",
                       "MEN1", "MLH1", "MSH2", "NF1", "NF2", "NOTCH1", "NPM1", "NR4A3", "NUP98", "PALB2", "PML", "PTEN",
                       "RB1", "RUNX1", "SDHB", "SDHD", "SMARCA4", "SMARCB1", "SOCS1", "STK11", "SUFU", "SUZ12", "SYK",
                       "TCF3", "TNFAIP3", "TP53", "TSC1", "TSC2", "VHL", "WRN", "WT1")

# chromatin remodeling
interesting_genes <- c(interesting_genes, "BRDT", "CHD1", "CHD2", "CHD3", "CHD4", "ARID3A", "HDAC1", "HDAC2", "HELLS",
                       "CITED1", "ARID4A", "BRD2", "SMARCA1", "SMARCA2", "HLTF", "SMARCA4", "SMARCB1", "SMARCC1",
                       "SMARCC2", "SMARCD1", "SMARCD2", "SMARCD3", "SMARCE1", "TAPBP", "BRD3", "ARID1A", "SMARCA5", 
                       "HAT1", "HDAC3", 
                       "KAT2B",  # "PCAF", 
                       "BAZ1B", "CDYL", "HDAC9", "HDAC4", "HDAC6", "HDAC5", "CITED2", "ARID3B",
                       "ARID5A", "BRD8", "BAZ2A", "BAZ1A", "BRD4", "BRD1", "CHD5", "BRD7", "BAZ2B", "SMARCAL1", 
                       "HDAC7",  # "HDAC7A",
                       "ARID4B", "CHRAC1", "CHD7", "HDAC8", "ARID1B", "CHD8", "BRD9", "HDAC11", "CHD9", "HDAC10", 
                       "KAT8",  # "MYST1",
                       "ARID5B", "CHD6", "CITED4")

interesting_genes <- unique(interesting_genes)
```

### Figure 4b/c

```{r}

x <- all_pc_bph$atac_log2FoldChange
y <- all_pc_bph$rna_log2FoldChange
gn <- all_pc_bph$gene_name

xlim <- c(-2,2)
ylim <- c(-10, 10)

x[x < xlim[1]] <- xlim[1]
x[x > xlim[2]] <- xlim[2]
y[y < ylim[1]] <- ylim[1]
y[y > ylim[2]] <- ylim[2]

df <- data.frame(x,y,gene_name=gn)


x <- de_pc_bph$atac_log2FoldChange
y <- de_pc_bph$rna_log2FoldChange
gn <- de_pc_bph$gene_name

x[x < xlim[1]] <- xlim[1]
x[x > xlim[2]] <- xlim[2]
y[y < ylim[1]] <- ylim[1]
y[y > ylim[2]] <- ylim[2]

df_deg <- data.frame(x,y,gene_name=gn)
# df_deg$gene_name <- ifelse(df_deg$gene_name %in% interesting_genes, as.character(df_deg$gene_name), "")

df$de <- ifelse(all_pc_bph$gene_name %in% de_pc_bph$gene_name, "Yes", "No")
df$gene_name <- ifelse(df$gene_name %in% interesting_genes & df$de == "Yes", as.character(df$gene_name), "")
# df$color <- ifelse(df$de == 1, "red", "gray70")


fit1 <- lm( I(y) ~ 0 + x, data = df)
deg_fit1 <- lm( I(y) ~ 0 + x, data = df_deg)

r2_fit1 <- summary(fit1)$r.squared
r2_deg_fit1 <- summary(deg_fit1)$r.squared

p1 <- ggplot(df, aes(x,y, label = gene_name, color = factor(de))) +
  scale_color_manual(values = c("gray60", "red"), name = "Differential expression") +
  # scale_color_manual(name = "Differential expression", 
  #                    values = c("gray60", "red", "gray40",                   "firebrick"), 
  #                    labels = c("No",     "Yes", "Regression for all genes", "Regression for DE genes"),
  #                    guide = guide_legend(override.aes = list(
  #                      linetype = c(rep("blank", 2), "solid", "solid"),
  #                      shape = c(rep(16, 2), NA, NA)),
  #                      size = 2)) +
  guides(colour = guide_legend(override.aes = list(size = 2))) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  background_grid(major = "xy", minor = "none", colour.major = "gray90") +
  geom_point(size = 0.2) +
  geom_point(data = df_deg, mapping = aes(x,y, label = gene_name), size = 0.2, color = "red") +
  geom_text_repel(force = 3, segment.colour = "grey30") +
  geom_abline(slope = coef(fit1), intercept = 0, color = "gray40") +
  geom_abline(slope = coef(deg_fit1), intercept = 0, color = "firebrick") +
  annotate("text",x = 1.75, y = -8.5, size = 5, label = sprintf("R^2==%.2f", r2_fit1), parse = TRUE, color = "gray40") + 
  annotate("text",x = 1.75, y = -7.5, size = 5, label = sprintf("R^2==%.2f", r2_deg_fit1), parse = TRUE, color = "firebrick") + 
  # geom_smooth(method = "lm", color = "gray40", se = FALSE) + 
  # geom_smooth(data = df_deg, mapping = aes(x,y), method = "lm", color = "firebrick", se = FALSE) + 
  xlab("ATAC-seq log2 fold change") +
  ylab("RNA-seq log2 fold change") +
  ggtitle("BPH vs PC") +
  annotate("text", x = 1.75,  y = 9.5,  size = 5, 
           label = sprintf("%d/%d",                      
                           sum(de_pc_bph$atac_log2FoldChange > 0 & de_pc_bph$rna_log2FoldChange > 0),
                           sum(all_pc_bph$quadrant == 1))) +
  annotate("text", x = 1.75,  y = -9.5, size = 5, 
           label = sprintf("%d/%d", 
                           sum(de_pc_bph$atac_log2FoldChange > 0 & de_pc_bph$rna_log2FoldChange < 0),
                           sum(all_pc_bph$quadrant == 2))) +
  annotate("text", x = -1.75, y = -9.5, size = 5, 
           label = sprintf("%d/%d",                                             
                           sum(de_pc_bph$atac_log2FoldChange < 0 & de_pc_bph$rna_log2FoldChange < 0),
                           sum(all_pc_bph$quadrant == 3))) +
  annotate("text", x = -1.75, y = 9.5,  size = 5, 
           label = sprintf("%d/%d",
                           sum(de_pc_bph$atac_log2FoldChange < 0 & de_pc_bph$rna_log2FoldChange > 0),
                           sum(all_pc_bph$quadrant == 4))) +
  theme(legend.position = "bottom",
        legend.justification = "center",
        legend.text = element_text(size = 15))



x <- all_crpc_pc$atac_log2FoldChange
y <- all_crpc_pc$rna_log2FoldChange
gn <- all_crpc_pc$gene_name

xlim <- c(-2,2)
ylim <- c(-10, 10)

x[x < xlim[1]] <- xlim[1]
x[x > xlim[2]] <- xlim[2]
y[y < ylim[1]] <- ylim[1]
y[y > ylim[2]] <- ylim[2]

df <- data.frame(x,y,gene_name=gn)


x <- de_crpc_pc$atac_log2FoldChange
y <- de_crpc_pc$rna_log2FoldChange
gn <- de_crpc_pc$gene_name

x[x < xlim[1]] <- xlim[1]
x[x > xlim[2]] <- xlim[2]
y[y < ylim[1]] <- ylim[1]
y[y > ylim[2]] <- ylim[2]

df_deg <- data.frame(x,y,gene_name=gn)
# df_deg$gene_name <- ifelse(df_deg$gene_name %in% interesting_genes, as.character(df_deg$gene_name), "")

df$de <- ifelse(all_crpc_pc$gene_name %in% de_crpc_pc$gene_name, "Yes", "No")
df$gene_name <- ifelse(df$gene_name %in% interesting_genes & df$de == "Yes", as.character(df$gene_name), "")
# df$color <- ifelse(df$de == 1, "red", "gray70")

fit2 <- lm( I(y) ~ 0 + x, data = df)
deg_fit2 <- lm( I(y) ~ 0 + x, data = df_deg)

r2_fit2 <- summary(fit2)$r.squared
r2_deg_fit2 <- summary(deg_fit2)$r.squared

p2 <- ggplot(df, aes(x, y, label = gene_name, color = factor(de))) +
  scale_color_manual(values = c("gray60", "red"), name = "Differential expression") +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  background_grid(major = "xy", 
                  minor = "none", 
                  colour.major = "gray90") +
  geom_point(size = 0.2) +
  geom_point(data = df_deg, 
             mapping = aes(x,y, label = gene_name), 
             size = 0.2, 
             color = "red") +
  geom_text_repel(force = 3, 
                  segment.colour = "grey30") +
  geom_abline(slope = coef(fit2), 
              intercept = 0, 
              color = "gray40") +
  geom_abline(slope = coef(deg_fit2), 
              intercept = 0, 
              color = "firebrick") +
  annotate("text",x = 1.75, y = -8.5, size = 5, label = sprintf("R^2==%.2f", r2_fit2), parse = TRUE, color = "gray40") + 
  annotate("text",x = 1.75, y = -7.5, size = 5, label = sprintf("R^2==%.2f", r2_deg_fit2), parse = TRUE, color = "firebrick") + 
  # geom_smooth(method = "lm", color = "gray40", se = FALSE) + 
  # geom_smooth(data = df_deg, mapping = aes(x,y), method = "lm", color = "firebrick", se = FALSE) + 
  xlab("ATAC-seq log2 fold change") +
  ylab("RNA-seq log2 fold change") +
  ggtitle("CRPC vs PC") +
  annotate("text", x = 1.75,  y = 9.5,  size = 5, 
           label = sprintf("%d/%d",                      
                           sum(de_crpc_pc$atac_log2FoldChange > 0 & de_crpc_pc$rna_log2FoldChange > 0),
                           sum(all_crpc_pc$quadrant == 1))) +
  annotate("text", x = 1.75,  y = -9.5, size = 5, 
           label = sprintf("%d/%d", 
                           sum(de_crpc_pc$atac_log2FoldChange > 0 & de_crpc_pc$rna_log2FoldChange < 0),
                           sum(all_crpc_pc$quadrant == 2))) +
  annotate("text", x = -1.75, y = -9.5, size = 5, 
           label = sprintf("%d/%d",                                             
                           sum(de_crpc_pc$atac_log2FoldChange < 0 & de_crpc_pc$rna_log2FoldChange < 0),
                           sum(all_crpc_pc$quadrant == 3))) +
  annotate("text", x = -1.75, y = 9.5,  size = 5, 
           label = sprintf("%d/%d",
                           sum(de_crpc_pc$atac_log2FoldChange < 0 & de_crpc_pc$rna_log2FoldChange > 0),
                           sum(all_crpc_pc$quadrant == 4))) +
  theme(legend.position = "bottom",
        legend.justification = "center",
        legend.text = element_text(size = 15))
  # theme(legend.position = "none")

# leg <- get_legend(p1)
# p1 <- p1 + theme(legend.position = "none")

gp <- get.gpar()
gp$fontsize <- 16
gp$fontface <- 2

p <- plot_grid(p1,p2, ncol = 2, align = "h")
# p <- plot_grid(textGrob("TSS", gp = gp), p, leg, nrow = 3, rel_heights = c(0.1, 1, 0.1))
# p <- plot_grid(textGrob("TSS", gp = gp), p, nrow = 2, rel_heights = c(0.1, 1))

ggsave(p, filename = "~/test/plot.pdf", width = 14, height = 7)


```

### Figure 4d

```{r}
library(MASS)

# interesting_genes <- c("AR","FOXA1",
#                        "NR4A3","CITED2",
#                        "JUN","MYC",
#                        "PIM1","BCL11A",
#                        "CCND2","KIT",
#                        "FGFR2","TNFAIP3")

interesting_genes <- c("AR", "BCL11A", "MYC")


gp <- get.gpar()
gp$fontsize <- 16
gp$fontface <- 2


plist <- setNames(vector("list", length(interesting_genes)), interesting_genes)
legend <- NULL
for (i in seq_along(interesting_genes)) {
  ensgid <- ann[ann$gene_name == interesting_genes[i] & ann$type == "gene"]$gene_id
  
  # if (ensgid %in% as.character(de_pc_bph$gene_id)) {
  #   pears1 <- de_pc_bph[as.character(de_pc_bph$gene_id) == ensgid, "pearson"] 
  #   spear1 <- de_pc_bph[as.character(de_pc_bph$gene_id) == ensgid, "spearman"]  
  # } else {
  #   pears1 <- NA
  #   spear1 <- NA
  # }
  # 
  # if (ensgid %in% as.character(de_crpc_pc$gene_id)) {
  #   pears2 <- de_crpc_pc[as.character(de_crpc_pc$gene_id) == ensgid, "pearson"]
  #   spear2 <- de_crpc_pc[as.character(de_crpc_pc$gene_id) == ensgid, "spearman"]
  # } else {
  #   pears2 <- NA
  #   spear2 <- NA
  # }
  # pearson <- c(pears1, pears2)
  # spearman <- c(spear1, spear2)
  # 
  # cordf <- data.frame(row.names = c("PC vs BPH", "CRPC vs PC"), 
  #                  pearson = pearson, 
  #                  spearman = spearman)
  # pt <- tableGrob(cordf)
  
  
  gexp <- assay(rna_dds,"norm.counts")[ensgid,]
  idx <- names(atac_rc_mat) %in% names(gexp)
  atac <- atac_rc_mat[gid == ensgid, ..idx]
  atac <- setNames(as.numeric(atac), names(atac))
  
  df <- as.data.frame(cbind(gexp, atac))
  df$type <- gsub("^(PC|BPH|CRPC)_.+", "\\1", rownames(df))
  df <- base::within(df, type <- factor(type, levels = c("BPH", "PC", "CRPC")))
  
  p <- ggplot(df, aes(gexp, atac)) + 
    coord_cartesian() +
    scale_color_manual(values = c("#33CC00", "#0066FF", "#FF0033"), name = "Sample\ntype") +
    geom_point(aes(color = type)) +
    geom_smooth(method = MASS::rlm, method.args = list(psi = psi.bisquare), se = FALSE, size = 1, colour = "gray60") +
    xlab("Gene expression") +
    ylab("Chromatin accessibility") +
    theme(aspect.ratio = 1)
          # axis.title = element_text(size = 10))
  
  # g <- plot_grid(pt, p, ncol = 2, rel_widths = c(1,1))
  
  # title <- textGrob(interesting_genes[i], gp = gp)
  # 
  # plist[[i]] <- plot_grid(title, g, nrow =2, rel_heights = c(0.1,1))
  
  if(is.null(legend)) {
    legend <- get_legend(p)
  }
  
  p <- p + theme(legend.position = "none")
  title <- textGrob(interesting_genes[i], gp = gp)
  plist[[i]] <- plot_grid(title, p, nrow =2, rel_heights = c(0.1,1))

}

pp <- plot_grid(plotlist = plist, ncol = 3)
pp <- plot_grid(pp, legend, ncol = 2, rel_widths = c(1,0.1))

# marrangeGrob(grobs = plist, ncol = 2, nrow = 2)
ggsave(pp, filename = "~/projects/atacseq_tamperepc/interesting_genes_scatter2.pdf",
       height = 5, width= 14)


```






