---
title: 'Tampere PC: correlation of ATAC signal at peak locations and gene expression within transcriptionally associated domains'
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Francesco Tabaro"
output:
  pdf_document:
  toc: true
toc_depth: 2
number_sections: true  
df_print: tibble
fig_caption: true
html_document:
  toc: true
toc_depth: 2
number_sections: true
fig_caption: true
df_print: paged
toc_float:
  collapsed: false
smooth_scroll: false
---

```{r setup}
knitr::opts_knit$set(root.dir="/bmt-data/genomics//projects//atac_workdir/")

library(GenomicAlignments)
library(GenomicInteractions)
library(data.table)
library(DESeq2)
library(openxlsx)

NCORES <- floor(parallel::detectCores() * 1)
setDTthreads(NCORES)
```

# Load data

## Precomputed correlation matrices

For computation method see `peaksVsGexp_ALLvALL.R`

```{r}

convertTable <- function (df) {
  df$peak_id <- as.character(df$peak_id)
  df$gene_id <- as.character(df$gene_id)
  return(df)
}

load("R_data/peaksVSgexp_lncap_tad_cormat.RData")
lncap_cormat <- convertTable(cormat)
dim(lncap_cormat)

load("R_data/peaksVSgexp_h1esc_tad_cormat.RData")
h1esc_cormat <- convertTable(cormat)
dim(h1esc_cormat)

```

### Overview plot

```{r}

doHist <- function(vect, main, label) {
  med <- median(vect,na.rm=T)
  sd <- sd(vect,na.rm=T)
  
  hist(vect, breaks="fd", col="grey80", 
       main = main,
       xlab = sprintf("%s correlation; median = %.2f; sd = %.2f", label, med, sd),
       sub = sprintf("N=%d", length(vect)))
  abline(v=med, lwd=2, col="dodgerblue")
}

par(mfrow=c(2,2))
doHist(lncap_cormat$pearson, "LnCAP", "Pearson")
doHist(lncap_cormat$spearman, "LnCAP", "Spearman")
doHist(h1esc_cormat$pearson, "H1 ESC", "Pearson")
doHist(h1esc_cormat$spearman, "H1 ESC", "Spearman")
```

```{r}
L <- list(lncap_pearson=lncap_cormat$pearson,
          lncap_pearson=lncap_cormat$spearman,
          h1esc_pearson=h1esc_cormat$pearson,
          h1esc_pearson=h1esc_cormat$spearman)
boxplot(L)

```

## Load Ensembl annotation
```{r}
source("R/getAnnotationTable.R")
ann <- getAnnotationTable()
```

## Load mirBase annotation
```{r}
mirbase <- import("/bmt-data/genomics/reference/mirbase_22/hsa.gff3")
seqinfo(mirbase) <- seqinfo(ann)
```


# Filter for correlating interactions

```{r}

COR_THR <- 0.5

dt <- as.data.table(lncap_cormat)
dt$gene_name <- ann$gene_name[match(dt$gene_id,ann$gene_id)]
dt[is.na(gene_name)] <- dt[is.na(gene_name), gene_name := gene_id]

# dt[abs(pearson) > COR_THR | abs(spearman) > COR_THR | abs(protein_pearson) > COR_THR | abs(protein_spearman) > COR_THR]

filterGenes <- function(gene_id, gene_name, pearson, spearman) {
  gkeep <- abs(pearson) > COR_THR | abs(spearman) > COR_THR
  paste(sprintf("%s (%s, pearson: %.2f, spearman: %.2f)", 
                gene_name[gkeep], gene_id[gkeep], 
                pearson[gkeep], spearman[gkeep]), 
        collapse = "; ")
}

filterProteins <- function (gene_id, gene_name, protein_pearson, protein_spearman) {
  pkeep <- !(is.na(protein_spearman) & is.na(protein_pearson)) & (abs(protein_spearman) > COR_THR | abs(protein_pearson) > COR_THR)
  paste(sprintf("%s (%s, pearson: %.2f, spearman: %.2f)", 
                gene_name[pkeep], gene_id[pkeep], 
                protein_pearson[pkeep], protein_spearman[pkeep]), 
        collapse = "; ")
}

aggregated <- dt[,.(ngenes = sum(abs(pearson) > COR_THR | abs(spearman) > COR_THR, na.rm = TRUE),
                    nproteins = sum(abs(protein_pearson) > COR_THR | abs(protein_spearman) > COR_THR, na.rm = TRUE),
                    genes = filterGenes(gene_id, gene_name, pearson, spearman),
                    proteins = filterProteins(gene_id, gene_name, protein_pearson, protein_spearman)), 
                 by = .(peak_id, tad_id)]

```


```{r}
COR_THR <- 0.5

lncap_cormat_all <- lncap_cormat
lncap_cormat <- subset(lncap_cormat, subset = abs(pearson) > COR_THR | abs(spearman) > COR_THR) # retain ~2%
dim(lncap_cormat)

h1esc_cormat <- subset(h1esc_cormat, subset = abs(pearson) > COR_THR | abs(spearman) > COR_THR) # retain ~1.8%
dim(h1esc_cormat)
```

```{r}
cat("Unique genes:",length(unique(lncap_cormat$gene_id)),"\n")
cat("Unique tad:",length(unique(lncap_cormat$tad_id)),"\n")
cat("Unique peaks:",length(unique(lncap_cormat$peak_id)),"\n")

tmp <- as.data.table(lncap_cormat)
tmpagg <- tmp[,.N,by=c("peak_id", "gene_id")]
cat("Unique peak-gene correlations:", nrow(tmpagg))
```

# Generate junction.bed files

```{r, eval=FALSE}
dumpJunctionsFile <- function(dat) {
  res <- mclapply(seq(nrow(dat)), function (i) {
    peak_id <- dat[i, "peak_id"]
    peak_range <- makeGRangesFromDataFrame(data.frame(chr   = gsub("(chr.*)_[0-9]+_[0-9]+", "\\1", peak_id),
                                                      start = gsub("chr.*_([0-9]+)_[0-9]+", "\\1", peak_id),
                                                      end   = gsub("chr.*_[0-9]+_([0-9]+)", "\\1", peak_id)),
                                           seqinfo = seqinfo(ann))
    genes <- dat[i, "gene_id"]
    if (genes %in% ann$gene_id) {
      gene_ranges <- granges(ann[ann$gene_id == genes & ann$type == "gene"])
      gene_ranges$name <- ann$gene_name[ann$gene_id==genes & ann$type=="gene"]
    } else if (genes %in% mirbase$Name) {
      gene_ranges <- granges(mirbase[mirbase$Name == genes & seqnames(mirbase) == seqnames(peak_range)])
      if(length(gene_ranges) > 1) {
        grdist <- distance(gene_ranges, peak_range) # takes into account strandness
        gene_ranges <- gene_ranges[which.min(grdist)]
      }
      gene_ranges$name <- genes
    } else {
      stop("Could not find gene id.")
    }
    pearsons <- as.numeric(dat[i,"pearson"])
    spearmans <- as.numeric(dat[i,"spearman"])
    gene_ranges$pearson <- pearsons[1:length(gene_ranges)] # some mir are duplicate. let's just pick the first one...
    gene_ranges$spearman <- spearmans[1:length(gene_ranges)] # some mir are duplicate. let's just pick the first one...
    peak_range <- rep(peak_range,length(gene_ranges))
    names(peak_range) <- as.character(peak_range)

    GenomicInteractions(GInteractions(peak_range, gene_ranges))
  }, mc.cores = NCORES)

  out <- do.call(c,res)

  df <- data.frame(chrom=seqnames(anchorOne(out)),
                   start=start(anchorOne(out)) + (width(anchorOne(out)) - 1) / 2,
                   end=start(anchorTwo(out)),
                   # gene_id=mcols(out)$anchor2.gene_id,
                   name=sprintf("%s->%s;pears=%.2f;spear=%.2f", as.character(anchorOne(out)), mcols(out)$anchor2.name, mcols(out)$anchor2.pearson, mcols(out)$anchor2.spearman),
                   # score=round((500 - 1) * (abs(mcols(out)$anchor2.pearson) - min(abs(mcols(out)$anchor2.pearson), na.rm=TRUE)) / (max(abs(mcols(out)$anchor2.pearson),na.rm=TRUE) - min(abs(mcols(out)$anchor2.pearson), na.rm=TRUE)),0),
                   score=0,
                   strand=ifelse(mcols(out)$anchor2.pearson > 0, "+", "-"),
                   thickStart=start(anchorOne(out)) + (width(anchorOne(out)) - 1) / 2,
                   thickEnd=start(anchorTwo(out)),
                   itemRGB=ifelse(mcols(out)$anchor2.pearson > 0, "255,0,0", "0,0,255"),
                   blockSizes="0",
                   blockStart="0")
  
  for (i in seq(nrow(df))) {
    if (df[i,"start"] > df[i,"end"]) {
      tmp <- df[i,"start"]
      df[i,"start"] <- df[i,"end"]
      df[i,"end"] <- tmp
    }
  }

  # write.table(df, file = file.path(dirname(fp), sprintf("%s_junctions.bed", sheet_names[sheet_number])), sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
  return(df)
}
t1 <- dumpJunctionsFile(lncap_cormat)
write.table(t1, file = "/bmt-data/genomics/projects/atac_workdir/peaks_and_tad/lncap/lncap_no_score_junctions.bed", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

#t2 <- dumpJunctionsFile(h1esc_cormat)
#write.table(t2, file = "/bmt-data/genomics/projects/atac_workdir/peaks_and_tad/h1_esc/h1esc_junctions_no_score.bed", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

### Generate junctions for mutual exclusive peaks

```{r, eval=FALSE}
mutual_pc <- fread("/bmt-data/genomics/projects/atac_workdir/mutual_exclusivity/bed/junction/pc_only_peaks.bed")
mutual_pc <- mutual_pc[,.(peak_id=paste(V1,V2,V3,sep="_"),V1=NULL,V2=NULL,V3=NULL)]
dim(mutual_pc)

mutual_crpc <- fread("/bmt-data/genomics/projects/atac_workdir/mutual_exclusivity/bed/junction/crpc_only_peaks.bed")
mutual_crpc <- mutual_crpc[,.(peak_id=paste(V1,V2,V3,sep="_"),V1=NULL,V2=NULL,V3=NULL)]
dim(mutual_crpc)

pc_subset <- merge(mutual_pc, lncap_cormat_all, by="peak_id")
dim(pc_subset)

crpc_subset <- merge(mutual_crpc, lncap_cormat_all, by="peak_id")
dim(crpc_subset)

# t4 <- dumpJunctionsFile(as.data.frame(pc_subset))
# write.table(t4, file = "/bmt-data/genomics/projects/atac_workdir/mutual_exclusivity/bed/junction/pc_only_peaks_all_junctions.bed", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
# 
# t5 <- dumpJunctionsFile(as.data.frame(crpc_subset))
# write.table(t5, file = "/bmt-data/genomics/projects/atac_workdir/mutual_exclusivity/bed/junction/crpc_only_peaks_all_junctions.bed", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

thr <- c(0.3,0.4,0.5)

for (THR in thr){
  t2 <- dumpJunctionsFile(as.data.frame(subset(pc_subset, abs(pearson)>THR | abs(spearman)>THR)))
  write.table(t2, file = sprintf("/bmt-data/genomics/projects/atac_workdir/mutual_exclusivity/bed/junction/pc_only_peaks_correlation_above_%s_junctions.bed", gsub("\\.","",THR)), sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
  
  t3 <- dumpJunctionsFile(as.data.frame(subset(crpc_subset, abs(pearson)>THR | abs(spearman)>THR)))
  write.table(t3, file = sprintf("/bmt-data/genomics/projects/atac_workdir/mutual_exclusivity/bed/junction/crpc_only_peaks_correlation_above_%s_junctions.bed", gsub("\\.","",THR)), sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
}
```

# Dump tables 

```{r}
library(openxlsx)
```

```{r}
gtrd_lncap <- import("/bmt-data/genomics/reference/gtrd/gtrd_meta_clusters_prostate_w_names.bed")
gtrd_all <- import("/bmt-data/genomics/reference/gtrd/gtrd_meta_clusters_all_w_names.bed")
```

```{r}
mergeGtrd <- function (dt, gtrd) {
  if ("gene_id" %in% names(dt)){
    dt[,tmpid := paste0(peak_id, gene_id)]
  }else{
    dt[,tmpid := peak_id]
  }
  gtrd_colname <- gsub("_", " ", deparse(substitute(gtrd)))
  df <- data.frame(chr   = gsub(pattern = "(chr.*)_[0-9]+_[0-9]+", replacement = "\\1", x = dt[["peak_id"]]),
                   start = gsub(pattern = "chr.*_([0-9]+)_[0-9]+", replacement = "\\1", x = dt[["peak_id"]]),
                   end   = gsub(pattern = "chr.*_[0-9]+_([0-9]+)", replacement = "\\1", x = dt[["peak_id"]]),
                   #name  = dt[["peak_id"]], 
                   tmpid   = dt[["tmpid"]], 
                   stringsAsFactors = FALSE)
  gr    <- makeGRangesFromDataFrame(df, starts.in.df.are.0based = FALSE, keep.extra.columns = TRUE)
  olaps <- findOverlaps(gtrd, gr, type = "within")
  tf_df <- data.table(tmpid = gr$tmpid[subjectHits(olaps)], gtrd_tf = gtrd$name[queryHits(olaps)])
  tf_df <- tf_df[, paste(unique(gtrd_tf), collapse = ", "), by = tmpid]
  names(tf_df) <- c("tmpid", gtrd_colname)
  ret <- merge(dt, tf_df, by = "tmpid",  all.x = TRUE)
  ret[,tmpid:=NULL]
  ret
}
```

```{r}
lncap_cormat <- mergeGtrd(as.data.table(lncap_cormat), gtrd_lncap)
lncap_cormat <- mergeGtrd(lncap_cormat, gtrd_all)

lncap_cormat_all <- mergeGtrd(as.data.table(lncap_cormat_all), gtrd_lncap)
lncap_cormat_all <- mergeGtrd(lncap_cormat_all, gtrd_all)

dim(lncap_cormat)
lncap_cormat <- merge(lncap_cormat, 
                      subset(as.data.frame(ann), subset = type == "gene", select = c("gene_id", "gene_name")), 
                      by = "gene_id", all.x = TRUE)

lncap_cormat_all <- merge(lncap_cormat_all, 
                          subset(as.data.frame(ann), subset = type == "gene", select = c("gene_id", "gene_name")), 
                          by = "gene_id", all.x = TRUE)


dim(lncap_cormat)

write.xlsx(lncap_cormat, file="tables/peaks_in_tad_cor05.xlsx")
write.xlsx(lncap_cormat_all, file="tables/peaks_in_tad_all.xlsx")
```


```{r}
aggregated <- mergeGtrd(aggregated, gtrd_lncap)
aggregated <- mergeGtrd(aggregated, gtrd_all)

write.xlsx(aggregated, file="tables/peaks_in_tad_aggregated_cor05.xlsx")
```

```{r}

lncap_cormat <- readWorkbook("tables/peaks_in_tad_cor05.xlsx", sheet = 1)
lncap_cormat <- as.data.table(lncap_cormat)

peaks_and_samples <- fread("heterogeneity/peaks_and_their_corresponding_samples.tsv")
peaks_and_samples[,`:=`(peak_id=paste(V1,V2,V3,sep="_"),V1=NULL,V2=NULL,V3=NULL)]

merged <- merge(lncap_cormat, peaks_and_samples, by="peak_id", all.x = T)

doGrepping <- function (pattern, str) {
  vect <- unlist(strsplit(str, ", "))
  any(grepl(sprintf("^%s_.+$", pattern), vect))
}

merged[,`:=`(peak_in_BPH  = mapply(doGrepping, V4, pattern = "BPH"),
             peak_in_PC   = mapply(doGrepping, V4, pattern = "PC"),
             peak_in_CRPC = mapply(doGrepping, V4, pattern = "CRPC"))]

write.xlsx(merged, file="tables/peaks_in_tad_aggregated_cor05_with_samples.xlsx")

```

# Plot correlation

```{r}
fp <- "tables/peaks_in_tad_cor05.xlsx"
dat <- readWorkbook(fp, sheet = 1)

doPlot <- function(vect, label) {
  
  signal <- hist(abs(vect), 10, plot = FALSE)
  signal$density <- signal$counts/sum(signal$counts)*100
  
  lo_signal <- loess.smooth(seq(0,1, length.out = length(signal$density)), signal$density, evaluation = 100) 
  
  plot(signal,
       freq=FALSE, 
       col = rgb(1,0,0,0.4), 
       ylab = "Relative frequency (%)", 
       xlim = c(0,1), 
       main = sprintf("Distribution of %s correlation coefficients (absolute values)", label),
       xlab = sprintf("%s correlation coefficient (absolute)", label))

  lines(lo_signal$x, lo_signal$y, lwd = 2, col = "red")
  # legend("topright", legend = c("Background", "Signal"), col = c("black", "red"), lwd = 2)

}

pdf("pictures/peaks_in_tad_hist_cor05.pdf")
par(mfrow=c(2,1))
doPlot(dat$spearman, "Spearman")
doPlot(dat$pearson, "Pearson")
dev.off()

```


# Plot number of peaks by TF

```{r, eval=FALSE}
countTf <- function(highcor){
  
  highcor <- highcor[!duplicated(highcor)]
  
  tfn <- unique(gtrd_all$name)
  pca_tf <- matrix(0, length(tfn), 3, dimnames = list(tfn, c("bph", "pc", "crpc")))
  all_tf <- matrix(0, length(tfn), 3, dimnames = list(tfn, c("bph", "pc", "crpc")))
  
  
  for(i in seq(nrow(highcor))) {
    samples <- unlist(strsplit(as.character(highcor[i,"samples"]), split = ", "))
    sample_type <- unique(tolower(gsub("^(PC|BPH|CRPC).*", "\\1", samples)))
    pca_ith_tf <- table(unlist(strsplit(as.character(highcor[i, "gtrd pca"]), ", ")))
    all_ith_tf <- table(unlist(strsplit(as.character(highcor[i, "gtrd all"]), ", ")))
    stopifnot(all(pca_ith_tf == 1) & all(all_ith_tf == 1))
    
    pca_ith_tf <- matrix(rep(pca_ith_tf, times = length(sample_type)), length(pca_ith_tf), length(sample_type), dimnames = list(names(pca_ith_tf), sample_type))
    all_ith_tf <- matrix(rep(all_ith_tf, times = length(sample_type)), length(all_ith_tf), length(sample_type), dimnames = list(names(all_ith_tf), sample_type))
    
    pca_tf[rownames(pca_ith_tf), sample_type] <- pca_tf[rownames(pca_ith_tf), sample_type] + pca_ith_tf  
    all_tf[rownames(all_ith_tf), sample_type] <- all_tf[rownames(all_ith_tf), sample_type] + all_ith_tf
    
  }
  
  tf_count_by_type <- merge(all_tf, pca_tf, by = 0, suffixes = c("_all", "_pca"))
  tf_count_by_type <- tf_count_by_type[!apply(tf_count_by_type[,-1],1,function(x) all(x == 0)),]
  
  return(tf_count_by_type)
}

sortCountFactor <- function (tf_count_by_type) {
  base::within(tf_count_by_type, {
    Row.names <- factor(Row.names, 
                        levels = Row.names[order(apply(tf_count_by_type[-1],1,function(x) sum(x)), decreasing = TRUE)])
  })
}

doPlot <- function (tf_count_by_type, plot_all = FALSE, selected = NULL) {
  
  if (!is.null(selected)) {
    tf_count_by_type <- tf_count_by_type[tf_count_by_type$Row.names %in% selected, ]
    tf_count_by_type <- sortCountFactor(tf_count_by_type)
  } else {
    tf_count_by_type <- sortCountFactor(tf_count_by_type)
  }
  
  if (!plot_all) {
    tf_count_by_type <- subset(tf_count_by_type, Row.names %in% levels(Row.names)[1:40])
  }
  
  m <- melt(tf_count_by_type)
  m$group <- gsub(".*_(pca|all)", "\\1", x = m$variable)
  m$variable <- gsub("^(bph|pc|crpc).*", "\\1", m$variable)
  
  sample_type <- c(bph="BPH", pc="PC", crpc="CRPC")
  gtrd_names <- c(pca="GTRD PCA", all="GTRD ALL")
  
  p <- ggplot(m, aes(x=as.factor(Row.names), y=value, fill=variable)) + 
    geom_bar(stat="identity", position="dodge") + 
    facet_grid(group ~ variable, labeller = labeller(variable=sample_type, group=gtrd_names)) +
    xlab("Transcription factor") +
    ylab("Number of peaks") + 
    ggtitle("Which TFBS are there in peaks correlating with gene expression within TADs?") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle=60, hjust=1, size = 6)) +
    scale_fill_manual(values = c(bph="#33CC00", pc="#0066FF", crpc="#FF0033"), name = "Sample group")
  return(p)
}

peaks_and_samples <- fread("heterogeneity/peaks_and_their_corresponding_samples.tsv")
peaks_and_samples[,`:=`(peak_id=paste(V1,V2,V3,sep="_"),V1=NULL,V2=NULL,V3=NULL)]
dim(peaks_and_samples)

highcor <- merge(lncap_cormat, peaks_and_samples, by="peak_id")
highcor <- highcor[,.(`gtrd pca` = `gtrd lncap`, `gtrd all`, samples = V4), by = "peak_id"]
dim(highcor)

tf_count <- countTf(highcor)

p <- doPlot(tf_count)
ggsave(p, filename = "/bmt-data/genomics/projects/atac_workdir/pictures/peaks_in_TAD_by_TF_v2.pdf", height = 5, width = 15)

p <- doPlot(tf_count, plot_all = TRUE)
ggsave(p, filename = "/bmt-data/genomics/projects/atac_workdir/pictures/peaks_in_TAD_by_TF_v2_allTF.pdf", height = 5, width = 200, limitsize = F)

# Plot TF from Figure 5
selected <- c("AR","FOXA1","ERG","HOXB13","NR3C1","GATA2","NKX3-1","MYC","CREB1","RUNX1","ESR1","CTCF","FOXP1","ASCL2","RELA","NANOG","E2F1","RUNX2","JUND","ETV1","GABPA","ELF1","ERF","GRHL2","TFAP4","VDR","ETS1","TCF7L2","TP63","ZFX","SFPQ","KDM5D","SMARCA5","ETV4","ELK4","BRD4","POU2F1","ETS2")
p <- doPlot(tf_count, selected = selected)
ggsave(p, filename = "/bmt-data/genomics/projects/atac_workdir/pictures/peaks_in_TAD_by_TF_v2_TFfromFig5.pdf", height = 5, width = 15)


# allcor <- merge(lncap_cormat_all, peaks_and_samples, by="peak_id")
# allcor <- allcor[,.(`gtrd pca` = `gtrd lncap`, `gtrd all` = `gtrd all`, samples = V4), by = peak_id]
# dim(highcor)
# p <- doPlot(allcor)
# ggsave(p, filename = "/bmt-data/genomics/projects/atac_workdir/pictures/all_peaks_in_TAD_by_TF_v2.pdf", height = 5, width = 15)
```

```{r}
computeGtrdBackground <- function () {
  all_tfbs_pca <- table(gtrd_lncap$name)
  all_tfbs_all <- table(gtrd_all$name)
  all_tfbs_mat <- matrix(0, length(all_tfbs_all), 2, dimnames = list(names(all_tfbs_all), c("gtrd_all", "gtrd_pca")))
  all_tfbs_mat[,1] <- all_tfbs_all
  all_tfbs_mat[match(names(all_tfbs_pca), rownames(all_tfbs_mat)), 2] <- all_tfbs_pca
  as.data.frame(all_tfbs_mat)
}

doPlotWithBackground <- function (tf_count, selected = NULL, freq = TRUE){
  if (!is.null(selected)){
    tf_count <- tf_count[tf_count$Row.names %in% selected,] #
    tf_count <- sortCountFactor(tf_count)
  } else {
    tf_count <- sortCountFactor(tf_count)
  }
  
  all_tfbs_df <- computeGtrdBackground()
  
  m <- melt(tf_count, value.name = "npeaks", variable.name = "sample_group")
  m <- merge(m, all_tfbs_df, all.x = TRUE, by.x = "Row.names", by.y = 0)
  m$group <- gsub(".*_(pca|all)", "\\1", x = m$sample_group)
  m$sample_group <- gsub("^(bph|pc|crpc).*", "\\1", m$sample_group)
  m2 <- melt(m)
  m2$variable <- factor(m2$variable, levels = c("gtrd_all", "gtrd_pca", "npeaks"))
  
  m2[m2$group == "pca" & m2$variable == "gtrd_all", "value"] <- NA
  m2[m2$group == "all" & m2$variable == "gtrd_pca", "value"] <- NA
  
  sample_type <- c(bph="BPH", pc="PC", crpc="CRPC")
  gtrd_names <- c(pca="GTRD PCA", all="GTRD ALL")
  
  if (freq){
    p <- ggplot(m2, aes(x = Row.names, y = value, fill = variable)) + 
      geom_bar(stat="identity", position = "fill" ) + 
      facet_grid(group ~ sample_group, labeller = labeller(sample_group=sample_type, group=gtrd_names)) +
      xlab("TF") + ylab("Relative frequency") +
      theme_minimal() +
      scale_fill_manual(values = c(gtrd_all="#FBB4AE", gtrd_pca="#B3CDE3", npeaks="#E41A1C"), name = "Peaks") +
      theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 6))
  } else {
    p <- ggplot(m2, aes(x = Row.names, y = value, fill = variable)) + 
      geom_bar(stat="identity") + 
      facet_grid(group ~ sample_group, labeller = labeller(sample_group=sample_type, group=gtrd_names)) +
      xlab("TF") + ylab("Frequency") +
      theme_minimal() +
      scale_fill_manual(values = c(gtrd_all="#FBB4AE", gtrd_pca="#B3CDE3", npeaks="#E41A1C"), name = "Peaks") +
      theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 6))
  }
  
  return(p)
}

p <- doPlotWithBackground(tf_count, selected = NULL, freq = FALSE)
ggsave(filename = "pictures/peaks_in_TAD_by_TF_countWrtBg_allTf.pdf", plot = p, width = 200, limitsize = FALSE)

p <- doPlotWithBackground(tf_count, selected = NULL)
ggsave(filename = "pictures/peaks_in_TAD_by_TF_freqWrtBg_allTf.pdf", plot = p, width = 200, limitsize = FALSE)

p <- doPlotWithBackground(tf_count, selected = selected, freq = FALSE)
ggsave(filename = "pictures/peaks_in_TAD_by_TF_countWrtBg_TFfromFig5.pdf", plot = p, width = 15)

p <- doPlotWithBackground(tf_count, selected = selected)
ggsave(filename = "pictures/peaks_in_TAD_by_TF_freqWrtBg_TFfromFig5.pdf", plot = p, width = 15)


```




```{r}
# dim(lncap_cormat_all)
# highcor <- merge(lncap_cormat_all, peaks_and_samples, by="peak_id")
# dim(highcor)
# 
# highcor <- highcor[,.(`gtrd pca`=unique(`gtrd lncap`),
#                       `gtrd all`=unique(`gtrd all`),
#                       samples=V4),
#                    by=peak_id]
# highcor <- highcor[!duplicated(highcor)]
# 
# tfn <- unique(gtrd_all$name)
# pca_tf <- matrix(0, length(tfn), 3, dimnames = list(tfn, c("bph", "pc", "crpc")))
# all_tf <- matrix(0, length(tfn), 3, dimnames = list(tfn, c("bph", "pc", "crpc")))
# 
# 
# for(i in seq(nrow(highcor))) {
#   samples <- unlist(strsplit(as.character(highcor[i,"samples"]), split = ", "))
#   sample_type <- unique(tolower(gsub("^(PC|BPH|CRPC).*", "\\1", samples)))
#   pca_ith_tf <- table(unlist(strsplit(as.character(highcor[i, "gtrd pca"]), ", ")))
#   all_ith_tf <- table(unlist(strsplit(as.character(highcor[i, "gtrd all"]), ", ")))
#   
#   pca_ith_tf <- matrix(rep(pca_ith_tf, times = length(sample_type)), length(pca_ith_tf), length(sample_type), dimnames = list(names(pca_ith_tf), sample_type))
#   all_ith_tf <- matrix(rep(all_ith_tf, times = length(sample_type)), length(all_ith_tf), length(sample_type), dimnames = list(names(all_ith_tf), sample_type))
#   
#   pca_tf[rownames(pca_ith_tf), sample_type] <- pca_tf[rownames(pca_ith_tf), sample_type] + pca_ith_tf  
#   all_tf[rownames(all_ith_tf), sample_type] <- all_tf[rownames(all_ith_tf), sample_type] + all_ith_tf
#   
# }
# 
# tf_count_by_type <- merge(all_tf, pca_tf, by = 0, suffixes = c("_all", "_pca"))
# tf_count_by_type <- tf_count_by_type[!apply(tf_count_by_type[,-1],1,function(x) all(x == 0)),]
# tf_count_by_type <- base::within(tf_count_by_type, Row.names <- factor(Row.names, levels = Row.names[order(apply(tf_count_by_type[-1],1,function(x) sum(x[4:6])), decreasing = TRUE)]))
# tf_count_by_type <- subset(tf_count_by_type, Row.names %in% levels(Row.names)[1:40])
# 
# m <- melt(tf_count_by_type)
# m$group <- gsub(".*_(pca|all)", "\\1", x = m$variable)
# m$variable <- gsub("^(bph|pc|crpc).*", "\\1", m$variable)
# 
# 
# p <- ggplot(m, aes(x=as.factor(Row.names), y=value, fill=variable)) + 
#   geom_bar(stat="identity", position="dodge") + 
#   facet_grid(~ group) +
#   theme(axis.text.x = element_text(angle=60, hjust=1, size = 6)) +
#   xlab("Transcription factor") + 
#   ylab("Occurrencies") +
#   ggtitle("GTRD TFBS in all peaks within LNCaP TAD.")
# 
# ggsave(p, filename = "/bmt-data/genomics/projects/atac_workdir/pictures/peaks_in_TAD_by_TF.pdf", height = 10, width = 10)
# print(p)
```



```{r}
agg <- lncap_cormat[, .(
  ngenes = .N,
  genes = paste(sprintf("%s->%s (pearson: %.2f, spearman: %.2f)", peak_id, gene_id, pearson, spearman), collapse = "; ")
), by = .(peak_id, tad_id)][ngenes > 1]
```

```{r}

```

# Do SPINK1 analysis

```{r}
sample_sheet <- read.csv("sample_sheet.csv", comment.char = "#", strip.white = T, stringsAsFactors = FALSE)
spink1pos <- subset(sample_sheet, Overexpr == "SPINK1", select = "Sample.name")$Sample.name
spink1 <- "ENSG00000164266"

sbst <- merge(subset(lncap_cormat_all, gene_id == spink1), peaks_and_samples, by="peak_id")
dim(sbst)

tfn <- unique(gtrd_all$name)
pca_tf_spink1 <- matrix(0, length(tfn), 3, dimnames = list(tfn, c("bph", "pc", "crpc")))
all_tf_spink1 <- matrix(0, length(tfn), 3, dimnames = list(tfn, c("bph", "pc", "crpc")))

for(i in seq(nrow(sbst))) {
  samples <- unlist(strsplit(as.character(sbst[i,"V4"]), split = ", "))
  sample_type <- unique(tolower(gsub("^(PC|BPH|CRPC).*", "\\1", samples)))
  pca_ith_tf <- table(unlist(strsplit(as.character(sbst[i, "gtrd lncap"]), ", ")))
  all_ith_tf <- table(unlist(strsplit(as.character(sbst[i, "gtrd all"]), ", ")))
  
  pca_ith_tf <- matrix(rep(pca_ith_tf, times = length(sample_type)), length(pca_ith_tf), length(sample_type), dimnames = list(names(pca_ith_tf), sample_type))
  all_ith_tf <- matrix(rep(all_ith_tf, times = length(sample_type)), length(all_ith_tf), length(sample_type), dimnames = list(names(all_ith_tf), sample_type))
  
  pca_tf_spink1[rownames(pca_ith_tf), sample_type] <- pca_tf_spink1[rownames(pca_ith_tf), sample_type] + pca_ith_tf  
  all_tf_spink1[rownames(all_ith_tf), sample_type] <- all_tf_spink1[rownames(all_ith_tf), sample_type] + all_ith_tf
}

tf_count_by_type <- merge(all_tf_spink1, pca_tf_spink1, by = 0, suffixes = c("_all", "_pca"))
tf_count_by_type <- tf_count_by_type[!apply(tf_count_by_type[,-1],1,function(x) all(x == 0)),]
tf_count_by_type <- base::within(tf_count_by_type, Row.names <- factor(Row.names, levels = Row.names[order(apply(tf_count_by_type[-1],1,function(x) sum(x[4:6])), decreasing = TRUE)]))
tf_count_by_type <- subset(tf_count_by_type, Row.names %in% levels(Row.names)[1:25])

m <- melt(tf_count_by_type)
m$group <- gsub(".*_(pca|all)", "\\1", x = m$variable)
m$variable <- gsub("^(bph|pc|crpc).*", "\\1", m$variable)


p <- ggplot(m, aes(x=as.factor(Row.names), y=value, fill=variable)) + 
  geom_bar(stat="identity", position="dodge") + 
  facet_grid(~ group) +
  theme(axis.text.x = element_text(angle=60, hjust=1, size = 6)) +
  xlab("Transcription factor") + 
  ylab("Occurrencies") +
  ggtitle("GTRD TFBS in all peaks within LNCaP TAD.")

print(p)

```

```{r}
peaks_and_samples <- fread("heterogeneity/peaks_and_their_corresponding_samples.tsv")
peaks_and_samples[,`:=`(peak_id=paste(V1,V2,V3,sep="_"),V1=NULL,V2=NULL,V3=NULL)]
dim(peaks_and_samples)

# spink1_pos_sample_sheet <- subset(sample_sheet, Overexpr == "SPINK1")
# spink1_pos_sample_sheet

sample_names <- sample_sheet[,"Sample.name"]
count <- setNames(do.call(c, mclapply(seq(length(sample_names)), function(i) {
  sum(apply(peaks_and_samples[,.(V4)],1, grepl, pattern = sample_names[i]))
}, mc.cores = NCORES)), sample_names)

spink1_status <- setNames(ifelse(sample_sheet[,"Overexpr"] == "SPINK1","POS","NEG"), sample_names)
df <- data.frame(count, spink1_status, sample = sample_names)
df <- subset(df, grepl("^PC|CRPC", sample))

ggplot(df,aes(reorder(sample, -count), count, fill = spink1_status)) + 
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=90, hjust=1)) +
  ggtitle("Number of unified peaks per sample.")

```
```{r}
library(cowplot)

spink1 <- "ENSG00000164266"
tad <- "chr5:147400001-147840000"

spink1_pos_sample_names <- subset(sample_sheet, Overexpr == "SPINK1", select = "Sample.name", drop = TRUE) # 5, 2 PC and 3 CRPC
spink1_neg_sample_names <- sample_sheet$Sample.name[grepl("^PC|CRPC", sample_sheet$Sample.name) & 
                                                      !sample_sheet$Sample.name %in% spink1_pos_sample_names & 
                                                      sample_sheet$Overexpr != "ERG"] # 9, 6 PC and 3 CRPC

spink1_pos_pattern <- paste0("^",spink1_pos_sample_names,"$")
spink1_neg_pattern <- paste0("^",spink1_neg_sample_names,"$")

all_tf <- unique(gtrd_all$name)
all_tf <- paste0("^", all_tf, "$")

sbst <- merge(subset(lncap_cormat_all, tad_id == tad & gene_id == spink1), peaks_and_samples, by="peak_id")
```

```{r}
res <- setNames(mclapply(seq_along(all_tf), function(i) {
  tf <- all_tf[i]
  tf <- paste0("^",tf,"$")

  spink1_pos_with_tfbs <- sum(sapply(sbst$`gtrd lncap`, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    any(grepl(tf, each_split))
  }) & sapply(sbst$V4, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    any(sapply(spink1_pos_pattern, grepl, x = each_split)) & !any(sapply(spink1_neg_pattern, grepl, x = each_split))
  }))
  
  spink1_pos_without_tfbs <- sum(sapply(sbst$`gtrd lncap`, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    !any(grepl(tf, each_split))
  }) & sapply(sbst$V4, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    any(sapply(spink1_pos_pattern, grepl, x = each_split)) & !any(sapply(spink1_neg_pattern, grepl, x = each_split))
  }))
  
  spink1_neg_with_tfbs <- sum(sapply(sbst$`gtrd lncap`, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    any(grepl(tf, each_split))
  }) & sapply(sbst$V4, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    !any(sapply(spink1_pos_pattern, grepl, x = each_split)) & any(sapply(spink1_neg_pattern, grepl, x = each_split))
  }))
  
  spink1_neg_without_tfbs <- sum(sapply(sbst$`gtrd lncap`, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    !any(grepl(tf, each_split))
  }) & sapply(sbst$V4, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    !any(sapply(spink1_pos_pattern, grepl, x = each_split)) & any(sapply(spink1_neg_pattern, grepl, x = each_split))
  }))

  # total_tfbs <- sum(grepl(tf, sbst$`gtrd lncap`))
  
  contmat <- matrix(c(spink1_pos_with_tfbs, spink1_pos_without_tfbs, spink1_neg_with_tfbs, spink1_neg_without_tfbs), 2,2)
  fisher <- fisher.test(contmat)
  
  return(list(fisher, contmat))
}), all_tf)

tmp <- lapply(res, function(x) {
  if (x[[1]]$p.value < 0.1) {
    return(x)
  } else {
    return(NULL)
  }
})
tmp[!sapply(tmp,is.null)]
```


```{r}
res <- setNames(mclapply(seq_along(all_tf), function(i) {
  tf <- all_tf[i]
  tf <- paste0("^",tf,"$")

  spink1_pos_with_tfbs <- sum(sapply(sbst$`gtrd all`, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    any(grepl(tf, each_split))
  }) & sapply(sbst$V4, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    any(sapply(spink1_pos_pattern, grepl, x = each_split)) & !any(sapply(spink1_neg_pattern, grepl, x = each_split))
  }))
  
  spink1_pos_without_tfbs <- sum(sapply(sbst$`gtrd all`, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    !any(grepl(tf, each_split))
  }) & sapply(sbst$V4, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    any(sapply(spink1_pos_pattern, grepl, x = each_split)) & !any(sapply(spink1_neg_pattern, grepl, x = each_split))
  }))
  
  spink1_neg_with_tfbs <- sum(sapply(sbst$`gtrd all`, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    any(grepl(tf, each_split))
  }) & sapply(sbst$V4, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    !any(sapply(spink1_pos_pattern, grepl, x = each_split)) & any(sapply(spink1_neg_pattern, grepl, x = each_split))
  }))
  
  spink1_neg_without_tfbs <- sum(sapply(sbst$`gtrd all`, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    !any(grepl(tf, each_split))
  }) & sapply(sbst$V4, function(each) {
    each_split <- unlist(strsplit(each, ", ", fixed = TRUE))
    !any(sapply(spink1_pos_pattern, grepl, x = each_split)) & any(sapply(spink1_neg_pattern, grepl, x = each_split))
  }))

  # total_tfbs <- sum(grepl(tf, sbst$`gtrd lncap`))
  
  contmat <- matrix(c(spink1_pos_with_tfbs, spink1_pos_without_tfbs, spink1_neg_with_tfbs, spink1_neg_without_tfbs), 2,2)
  fisher <- fisher.test(contmat)
  
  return(list(fisher, contmat))
}), all_tf)

pv <- p.adjust(sapply(res, function(x) x[[1]]$p.value), "fdr")
hist(pv,50, xlim=c(0,1)); abline(v=0.05,lwd=2,col="dodgerblue")


tmp <- lapply(res, function(x) {
  if (x[[1]]$p.value < 0.1) {
    return(x)
  } else {
    return(NULL)
  }
})
names(tmp[!sapply(tmp,is.null)])
```

# Check correlation with known oncogenes and tumor suppressor genes 

```{r}
# prostate cancer  
pca_related_genes <- c("KLK3",# aka PSA
                       "CDH1", "FOXA1", "CDKN1B", "ATM", "NKX3-1", "MED12", "PTEN", "RB1","AR")
interesting_genes <- pca_related_genes

# oncogenes (supp table 3 of http://cancerres.aacrjournals.org/content/canres/early/2012/01/17/0008-5472.CAN-11-2266.full.pdf)
general_oncogenes <- c("ABL1", "ABL2", "AKT1", "ATF1", "BCL11A", "BCL2", "BCL3", "BCL6", "BCR", "BRAF", "CARD11",
                       "CBLB", "CCND1", "CCND2", "CCND3", "CDX2", "CTNNB1", "DDB2", "DDIT3", "DDX6", "DEK", "EGFR",
                       "ELK4", "ERBB2", "ETV4", "ETV6", 
                       "MECOM",  #"EVI1", 
                       "EWSR1", "FEV", "FGFR1", "FGFR1OP", "FGFR2", "FUS", "GOLGA5", "GOPC", "HMGA1", "HMGA2", "HRAS",
                       "IRF4", "JUN", "KIT", "KRAS", "LCK", "LMO2", "MAF", "MAFB", "MAML2", "MDM2", "MET", "MITF", 
                       "KMT2A", # "MLL", 
                       "MPL", "MYB", "MYC", 
                       "MYCL", # "MYCL1", 
                       "MYCN", "NCOA4","NFKB2", "NRAS", "NTRK1", "NUP214", "PAX8", "PDGFB", "PIK3CA", "PIM1", "PLAG1",
                       "PPARG", "PTPN11","RAF1", "REL", "RET", "ROS1", "SMO", "SS18", "TCL1A", "TET2", "TFG", "TLX1",
                       "TPR", "USP6")
interesting_genes <- c(interesting_genes, general_oncogenes)

# tumor suppressor (http://cancerres.aacrjournals.org/content/canres/early/2012/01/17/0008-5472.CAN-11-2266.full.pdf)
general_tumor_suppressor <- c("APC", "ARHGEF12", "ATM", "BCL11B", "BLM", "BMPR1A", "BRCA1", "BRCA2", "CARS",
                       "CBFA2T3", "CDH1", "CDH11", "CDK6", "CDKN2C", "CEBPA", "CHEK2", "CREB1", "CREBBP", "CYLD", "DDX5",
                       "EXT1", "EXT2", "FBXW7", "FH", "FLT3", "FOXP1", "GPC3", "IDH1", "IL2", "JAK2", "MAP2K4", "MDM4",
                       "MEN1", "MLH1", "MSH2", "NF1", "NF2", "NOTCH1", "NPM1", "NR4A3", "NUP98", "PALB2", "PML", "PTEN",
                       "RB1", "RUNX1", "SDHB", "SDHD", "SMARCA4", "SMARCB1", "SOCS1", "STK11", "SUFU", "SUZ12", "SYK",
                       "TCF3", "TNFAIP3", "TP53", "TSC1", "TSC2", "VHL", "WRN", "WT1")
interesting_genes <- c(interesting_genes, general_tumor_suppressor)

# chromatin remodeling (https://www.qiagen.com/ca/resources/resourcedetail?id=9979518e-d773-43a4-973b-73c6eca28ee6&lang=en)
chromatin_remodelers <- c("BRDT", "CHD1", "CHD2", "CHD3", "CHD4", "ARID3A", "HDAC1", "HDAC2", "HELLS",
                       "CITED1", "ARID4A", "BRD2", "SMARCA1", "SMARCA2", "HLTF", "SMARCA4", "SMARCB1", "SMARCC1",
                       "SMARCC2", "SMARCD1", "SMARCD2", "SMARCD3", "SMARCE1", "TAPBP", "BRD3", "ARID1A", "SMARCA5", 
                       "HAT1", "HDAC3", 
                       "KAT2B",  # "PCAF", 
                       "BAZ1B", "CDYL", "HDAC9", "HDAC4", "HDAC6", "HDAC5", "CITED2", "ARID3B",
                       "ARID5A", "BRD8", "BAZ2A", "BAZ1A", "BRD4", "BRD1", "CHD5", "BRD7", "BAZ2B", "SMARCAL1", 
                       "HDAC7",  # "HDAC7A",
                       "ARID4B", "CHRAC1", "CHD7", "HDAC8", "ARID1B", "CHD8", "BRD9", "HDAC11", "CHD9", "HDAC10", 
                       "KAT8",  # "MYST1",
                       "ARID5B", "CHD6", "CITED4")
interesting_genes <- c(interesting_genes, chromatin_remodelers)

interesting_genes <- unique(interesting_genes)
```

```{r}
fn <- "tables/peaks_in_tad_cor05.xlsx"
dat <- readWorkbook(fn)

assignGeneClass <- function(n) {
  sapply(n, function(x) {
    if (x %in% pca_related_genes) {
      return("Prostate Cancer gene")
    } else if (x %in% general_oncogenes) {
      return("Oncogene")
    } else if (x %in% general_tumor_suppressor) {
      return("Tumor suppressor gene")
    } else if (x %in% chromatin_remodelers) {
      return("Chromatin remodeller")
    }
  })
}

df <- subset(dat, gene_name %in% interesting_genes)
df$gene_class <- assignGeneClass(df$gene_name)
table(df$gene_name,df$gene_class)

computeDistance <- function(gn){
  ar <- subset(df, gene_name == gn)
  gr1 <- makeGRangesFromDataFrame(data.frame(chr   = gsub("(chr.+)_[0-9]+_[0-9]+", "\\1", ar$peak_id),
                                             start = gsub("chr.+_([0-9]+)_[0-9]+", "\\1", ar$peak_id),
                                             end   = gsub("chr.+_[0-9]+_([0-9]+)", "\\1", ar$peak_id)))
  # rtracklayer::export(gr1,"~/test/test.bed",format = "BED")
  argr <- subset(ann, gene_name == gn & type == "gene")
  
  # follow_hits <- follow(gr1, argr, select="all")
  precede_hits <- precede(gr1, argr, select="all")
  distance <- distance(argr, gr1)
  
  distance[queryHits(precede_hits)] <- -1 * distance[queryHits(precede_hits)]
  return(distance)
}

gns <- unique(df$gene_name)
for (gn in gns) {
  distance <- computeDistance(gn)
  df[df$gene_name == gn, "distance"] <- distance  
}

# hist(distance)

length(unique(df[df$gene_class == "Oncogene", "gene_name"]))
length(unique(df[df$gene_class == "Tumor suppressor gene", "gene_name"]))
length(unique(df[df$gene_class == "Chromatin remodeller", "gene_name"]))

write.xlsx(df, "tables/peaks_in_tad_cor05_interesting_genes.xlsx")
```

# Compute backgroung distributions

```{r}
load("R_data/RNAdds.RData")
load("R_data/sRNA_dds.RData")
load("R_data/peaksVSgexp_lncap_tad_cormat.RData")

cormat <- as.data.table(cormat)
cormat$peak_id <- as.character(cormat$peak_id)
cormat$gene_id <- as.character(cormat$gene_id)

mir_annot <- mcols(rowRanges(srna_se_filter))
all_gr <- c(granges(rowRanges(rna_dds)), granges(rowRanges(srna_se_filter)))

peaks <- rtracklayer::import("peak_tss_dar_quantifications/peak_coordinates.bed")
start(peaks) <- start(peaks) - 1
names(peaks) <- paste(seqnames(peaks), start(peaks), end(peaks), sep= "_")

tads <- rtracklayer::import("dar_and_tad/lncap/lncap_tads.bed")
start(tads) <- start(tads) - 1

pairs <- mclapply(seq_along(tads), function (i) {
  tad <- tads[i]
  genes_inside <- c(rowRanges(rna_dds)$gene_id[rowRanges(rna_dds) %over% tad], 
                    rowRanges(srna_se_filter)$Name[rowRanges(srna_se_filter) %over% tad])
  peaks_inside <- names(peaks)[peaks %over% tad]
  return(expand.grid(genes_inside, peaks_inside))
}, mc.cores = NCORES)

pairs <- do.call(rbind, pairs)

write.table(pairs, 
            "peaks_and_tad/lncap/correlation_gene_expression/peaks_genes_within_tad_pairs.tsv", 
            row.names = FALSE,
            col.names = FALSE, 
            quote = FALSE)

# signal <- vector(mode = "numeric", nrow(pairs))
# for (i in seq(nrow(pairs))) {
#   peak <- pairs[i, "Var2"]
#   gene <- pairs[i, "Var1"]
#   signal[i] <- cormat[peak_id == peak & gene_id == gene, spearman]
#   cormat[peak_id == peak & gene_id == gene, spearman := NA]
#   
# }

# doPlot <- function(bg_vect, signal_vect, label) {
#   
#   bg <- hist(bg_vect, 30, plot = FALSE)
#   # bg$counts <- log10(bg$counts)
#   # bg$density <- bg$counts/sum(bg$counts)*100
#   
#   signal <- hist(signal_vect, 30, plot = FALSE)
#   # signal$counts <- log10(signal$counts)
#   # signal$density <- signal$counts/sum(signal$counts)*100
#   
#   # lo_bg <- loess.smooth(seq(-1,1, length.out = length(bg$density)), bg$density, evaluation = 100)
#   # lo_signal <- loess.smooth(seq(-1,1, length.out = length(signal$density)), signal$density, evaluation = 100) 
# 
#   plot(bg,
#        # freq=FALSE, 
#        xlim = c(-1,1),
#        ylab = "Frequency", 
#        main = sprintf("Distribution of %s correlation coefficients", label),
#        xlab = sprintf("%s correlation coefficient", label),
#        col = rgb(211/255,211/255,211/255,0.4))
#   
#   plot(signal,
#        add = TRUE,
#        # freq=FALSE, 
#        col = rgb(1,0,0,0.4))
#   
#   # lines(lo_bg$x, lo_bg$y, lwd = 2)
#   # lines(lo_signal$x, lo_signal$y, lwd = 2, col = "red")
#   legend("topright", legend = c("Background", "Signal"), col = c("black", "red"), lwd = 2)
# 
# }
# 
# pdf("pictures/peaks_in_tad_correlations_bg_and_signal_hist.pdf")
# # par(mfrow = c(2,1))
# doPlot(as.numeric(cormat[,spearman]), signal, "Spearman")
# # doPlot(all_pearson_cor, 
# #        as.numeric(gsub(".*pearson: (-?[0-9\\.]*), .*\\)", "\\1", unlist(strsplit(s$genes, "; ")))), 
# #        "Pearson")
# dev.off()


```

```{r}




```

