---
title: 'Tampere PC: correlation of ATAC signal at peak locations and gene expression of closest annotated gene'
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Francesco Tabaro"
output:
  pdf_document:
  toc: true
toc_depth: 2
number_sections: true  
df_print: tibble
fig_caption: true
html_document:
  toc: true
toc_depth: 2
number_sections: true
fig_caption: true
df_print: paged
toc_float:
  collapsed: false
smooth_scroll: false
---

```{r setup}
knitr::opts_knit$set(root.dir="/bmt-data/genomics//projects//atac_workdir/")

library(GenomicAlignments)
library(data.table)
library(DESeq2)

NCORES <- floor(parallel::detectCores() * 0.75)
setDTthreads(NCORES)
```

# Load all data

## Ensembl annotation

```{r}
source("R/getAnnotationTable.R")
ann <- getAnnotationTable()
ann.gene.gr <- ann[ann$type == "gene"]
ann.gene    <- as.data.frame(ann.gene.gr)
```


## RNA-seq

```{r}
load("R_data/RNAdds.RData")
```

## sRNA-seq

```{r}
load("R_data/sRNA_dds.RData")


mirna_map  <- data.table(mirna_name=rownames(srna_se_filter), mirna_id=rep(NA_character_,nrow(srna_se_filter)))
for(i in seq(nrow(srna_se_filter))) {
  pat <- gsub(pattern = "^hsa-mir-([^-]*).*$", replacement = "mir\\1", x = rownames(srna_se_filter)[i], ignore.case = TRUE)
  idx <- grep(pat, ann.gene$gene_name, ignore.case = TRUE)
  mirna_map[i, "mirna_id"] <- ifelse(length(idx)>0, ann.gene$gene_id[idx], NA_character_)
}
```

## Proteins
```{r}
uniprot <- fread("tables/swissprotdata_normalized.tsv")

biomart_map <- fread("tables/uniprot2ensembleGeneID/biomart_missing_uniprot.tsv", select = c(1,2)) # mapping from Biomart (Ensembl 90)
names(biomart_map) <- c("uniprot", "ensgid")
biomart_map[, ensgid := paste(ensgid, collapse = "|"), by = uniprot]

uniprot_map <- fread("tables/uniprot2ensembleGeneID/uniprot_mapping_missed.tsv") # mapping from Uniprot (unknown Ensembl)
names(uniprot_map) <- c("uniprot", "ensgid")
uniprot_map[, ensgid := paste(ensgid, collapse = "|"), by = uniprot]

missing_map <- rbind(biomart_map, uniprot_map)
missing_map <- missing_map[!duplicated(missing_map)]
missing_map[, ensgid := sapply(ensgid, function(x) ifelse(any(grepl(x, ann.gene$gene_id)), grep(x, ann.gene$gene_id, value = TRUE), ""))]
missing_map <- missing_map[ensgid != ""]

uniprot[, ensgid := ann.gene$gene_id[match(gn, ann.gene$gene_name)]]
uniprot[match(missing_map$uniprot, uniprot), ensgid := missing_map[match(uniprot, uniprot), ensgid]]

```


## ATAC-seq

```{r}
atac_rc <- fread("peak_tss_dar_quantifications/peak_quantification.tsv.gz")
atac_rc[,`:=`(peak_id=paste(chrom,start,end,sep="_"), chrom=NULL, start=NULL, end=NULL)]
```

## Homer annotation

```{r}
peaks_annot <- fread("slopped_summits/homer/slopped_summits.homer.txt")
names(peaks_annot)[1] <- "peak_id"
```

# Find common samples

```{r}
shared.samples <- Reduce(intersect, list(names(atac_rc), colnames(rna_dds), colnames(srna_se_filter))) # 28
 
shared.samples.proteins <- intersect(names(atac_rc), colnames(uniprot)) # 26
```

# Merge RNA matrices

```{r}
rnaseq  <- as.data.table(assay(rna_dds,"norm.counts")[,shared.samples], keep.rownames = TRUE)
names(rnaseq)[1] <- "gene_id"

srnaseq <- as.data.table(assay(srna_se_filter, "norm.counts")[,shared.samples], keep.rownames = TRUE)
srnaseq <- merge(mirna_map, srnaseq, by.x = "mirna_name", by.y = "rn", all.y = TRUE)
srnaseq[,mirna_name:=NULL]
names(srnaseq)[1] <- "gene_id"

rna_rc <- rbind(rnaseq, srnaseq)
rna_rc <- as.data.table(rna_rc, keep.rownames = TRUE)

```

# Do correlation

## Genes

```{r}
res <- do.call(rbind, mclapply(seq(nrow(peaks_annot)), function (i) {

  pid <- peaks_annot[i, peak_id]
  ogid <- peaks_annot[i, `Nearest Ensembl`]

  atac <- setNames(as.numeric(atac_rc[peak_id == pid, ..shared.samples]), shared.samples)
  
  if (ogid %in% rna_rc[["gene_id"]]) {
    gid <- ogid
    tmp <- rna_rc[gene_id == gid]
    
    pearson <- rep(NA, nrow(tmp))
    spearman <- rep(NA, nrow(tmp))

    for (ii in seq(nrow(tmp))) {
      rna <- setNames(as.numeric(tmp[ii, ..shared.samples]), shared.samples)
      
      stopifnot(all(names(rna) == names(atac)))

      pearson[ii] <- cor(atac, rna, method = "pearson", use = "na.or.complete")
      spearman[ii] <- cor(atac, rna, method = "spearman", use = "na.or.complete")
    }
    
    # gexp <- data.table(pid,gid, round(pearson,2), round(spearman,2))
    # gexp <- gexp[,`:=`(pearson=paste(as.character(pearson), collapse = ", "), 
    #                    spearman=paste(as.character(spearman), collapse = ", ")), by = .(pid, gid)]
    
  } else {
    gid <- NA_character_
    pearson <- NA_real_
    spearman <- NA_real_

  }
  
  data.table(pid, gid, pearson, spearman)  
}, mc.cores = NCORES))
```



## Proteins

```{r}
res_prot <- do.call(rbind, mclapply(seq(nrow(peaks_annot)), function (i) {
  
  pid <- peaks_annot[i, peak_id]
  ogid <- peaks_annot[i, `Nearest Ensembl`]

  atac <- setNames(as.numeric(atac_rc[peak_id == pid, ..shared.samples.proteins]), shared.samples.proteins)
  
  if(ogid %in% uniprot[["ensgid"]]) {
    gid <- ogid
    
    unp <- uniprot[ensgid == gid, uniprot]
    tmp <- uniprot[ensgid == gid]
    
    pearson_protein <- rep(NA, nrow(tmp))
    spearman_protein <- rep(NA, nrow(tmp))
    
    for (ii in seq(nrow(tmp))) {
      prot <- setNames(as.numeric(tmp[ii, ..shared.samples.proteins]), shared.samples.proteins)
      
      stopifnot(all(names(atac) == names(prot)))
      
      pearson_protein[ii] <- cor(atac, prot, method="pearson", use="na.or.complete")
      spearman_protein[ii] <- cor(atac, prot, method="spearman", use="na.or.complete")
    }
    
    # pexp <- data.table(pid, gid, unp, round(pearson_protein,2), round(spearman_protein,2))
    # pexp <- pexp[,`:=`(pearson_protein=paste(as.character(pearson_protein), collapse=", "), 
    #                    spearman_protein=paste(as.character(spearman_protein), collapse=", "), 
    #                    unp=paste(unp, collapse = ", ")), by = .(pid, gid)]
    
  } else {
    gid <- NA_character_
    unp <- NA_character_
    pearson_protein <- NA_real_
    spearman_protein <- NA_real_
    
    
  }

  data.table(pid, gid, unp, pearson_protein, spearman_protein)
}, mc.cores = NCORES))

```

## Join

```{r}
res <- res[!is.na(gid)]
res_prot <- res_prot[!is.na(gid)]

res_all <- merge(res, res_prot, by = c("gid", "pid"), all.x = TRUE)

annot <- merge(peaks_annot, res_all, by.x = "peak_id", by.y = "pid", all.x = TRUE)
```

## Plot distributions

```{r}

doHist <- function(vect, main, label) {
  med <- median(vect, na.rm = TRUE)
  sd <- sd(vect, na.rm = TRUE)
  
  hist(vect, breaks = "fd", xlab = sprintf("%s coeff; median = %.2f; sd = %.2f", label, med, sd),
       sub = sprintf("n = %d", sum(!is.na(vect))), main = main)
  abline(v = med, lwd = 2, col = "dodgerblue")
}

par(mfrow=c(2,2))
doHist(res_all[["pearson"]], "Genes", "Pearson")
doHist(res_all[["spearman"]], "Genes", "Spearman")
doHist(res_all[["pearson_protein"]], "Proteins", "Prot Pearson")
doHist(res_all[["spearman_protein"]], "Proteins", "Prot Spearman")


```
Get number of correlating peaks
```{r}
tmp <- res[abs(pearson) > 0.5 | abs(spearman) > 0.5]
dim(tmp)
```

```{r}
library(VennDiagram)
library(openxlsx)

COR_THR <- 0.5

correlating_genes <- res_all[abs(pearson) > COR_THR | abs(spearman) > COR_THR, gid]
correlating_prot <- res_all[abs(pearson_protein) > COR_THR | abs(spearman_protein) > COR_THR, gid]

venn.diagram(list(genes=correlating_genes, proteins=correlating_prot), filename = "~/test/vd.tiff")

part <- get.venn.partitions(list(genes=correlating_genes, proteins=correlating_prot))
genes_only <- subset(part, ..set.. == "(genes)∖(proteins)")
genes_gid <- genes_only$..values..$`3`
genes_gnames <- ann$gene_name[ann$gene_id %in% genes_gid & ann$type == "gene"]

prot_only <- subset(part, ..set.. == "(proteins)∖(genes)")
prot_gid <- prot_only$..values..$`2`
prot_gnames <- ann$gene_name[ann$gene_id %in% prot_gid & ann$type == "gene"]

intersection <- subset(part, ..set.. == "genes∩proteins")
intersection_gid <- intersection$..values..$`1`
intersection_gnames <- ann$gene_name[ann$gene_id %in% intersection_gid & ann$type == "gene"]

write.xlsx(list(genes_only=genes_gnames, proteins_only=prot_gnames, intersection=intersection_gnames), 
           file="tables/peaks_to_closest_gene_cor05_by_venn_partition.xlsx")
```

# Annotate TADs

```{r}
lncap_tad <- import("dar_and_tad/lncap/lncap_tads.bed")
h1esc_tad <- import("dar_and_tad/h1_esc/h1_esc_tads.bed")
```

```{r}
mergeTad <- function (dt, tad) {
  tad_colname <- gsub("_", " ", deparse(substitute(tad)))
  dt[,tmpid:=paste0(peak_id,gid,unp)]
  df <- data.frame(chr   = gsub(pattern = "(chr.*)_[0-9]+_[0-9]+", replacement = "\\1", x = dt[["peak_id"]]),
                   start = gsub(pattern = "chr.*_([0-9]+)_[0-9]+", replacement = "\\1", x = dt[["peak_id"]]),
                   end   = gsub(pattern = "chr.*_[0-9]+_([0-9]+)", replacement = "\\1", x = dt[["peak_id"]]),
                   #name  = dt[["peak_id"]], 
                   tmpid   = dt[["tmpid"]], 
                   stringsAsFactors = FALSE)  
  gr    <- makeGRangesFromDataFrame(df, starts.in.df.are.0based = FALSE, keep.extra.columns = TRUE)
  olaps <- findOverlaps(tad, gr)
  tad_df <- data.table(tmpid = gr$tmpid[subjectHits(olaps)], tad_id = as.character(tad)[queryHits(olaps)])
  tad_df <- tad_df[, paste(unique(tad_id), collapse = ", "), by = tmpid]
  names(tad_df) <- c("tmpid", tad_colname)
  
  ret <- merge(dt, tad_df, by = "tmpid",  all.x = TRUE)
  ret[,tmpid:=NULL]
  ret
}

annot <- mergeTad(annot, lncap_tad)
annot <- mergeTad(annot, h1esc_tad)
```

# Annotate GTRD

```{r}
gtrd_pca <- import("/bmt-data/genomics/reference/gtrd/gtrd_meta_clusters_prostate_w_names.bed")
gtrd_all <- import("/bmt-data/genomics/reference/gtrd/gtrd_meta_clusters_all_w_names.bed")
```

```{r}
mergeGtrd <- function (dt, gtrd) {
  dt[,tmpid := paste0(peak_id,gid,unp)]
  gtrd_colname <- gsub("_", " ", deparse(substitute(gtrd)))
  df <- data.frame(chr   = gsub(pattern = "(chr.*)_[0-9]+_[0-9]+", replacement = "\\1", x = dt[["peak_id"]]),
                   start = gsub(pattern = "chr.*_([0-9]+)_[0-9]+", replacement = "\\1", x = dt[["peak_id"]]),
                   end   = gsub(pattern = "chr.*_[0-9]+_([0-9]+)", replacement = "\\1", x = dt[["peak_id"]]),
                   #name  = dt[["peak_id"]], 
                   tmpid   = dt[["tmpid"]], 
                   stringsAsFactors = FALSE)
  gr    <- makeGRangesFromDataFrame(df, starts.in.df.are.0based = FALSE, keep.extra.columns = TRUE)
  olaps <- findOverlaps(gtrd, gr, type = "within")
  tf_df <- data.table(tmpid = gr$tmpid[subjectHits(olaps)], gtrd_tf = gtrd$name[queryHits(olaps)])
  tf_df <- tf_df[, paste(unique(gtrd_tf), collapse = ", "), by = tmpid]
  names(tf_df) <- c("tmpid", gtrd_colname)
  ret <- merge(dt, tf_df, by = "tmpid",  all.x = TRUE)
  ret[,tmpid:=NULL]
  ret
}

annot <- mergeGtrd(annot, gtrd_pca)
annot <- mergeGtrd(annot, gtrd_all)
```

```{r}
annot_hicor <- subset(annot, abs(pearson) > 0.5 | abs(spearman) > 0.5)

par(mfrow=c(2,2))
hist(annot_hicor$pearson, "fd")
hist(annot_hicor$spearman, "fd")
hist(annot_hicor$pearson_protein, "fd")
hist(annot_hicor$spearman_protein, "fd")


write.xlsx(annot_hicor, file="tables/genes_to_closest_peak_cor05.xlsx")

```





