---
title: 'Tampere PC: Differentially accessible regions vs gene expression correlation'
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Francesco Tabaro"
output:
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: true  
    df_print: tibble
    fig_caption: true
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
    fig_caption: true
    df_print: paged
    toc_float:
      collapsed: false
      smooth_scroll: false
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir="/bmt-data/genomics//projects//atac_workdir/")
NCORES <- parallel::detectCores()

library(rtracklayer)
library(DESeq2)
library(data.table)
library(cowplot)

source("/bmt-data/genomics//projects//atac_workdir/R/doCorr.R")

setDTthreads(NCORES)
```

# Generate annotated DAR sets

Using Homer and default annotation files.

```{bash, eval=FALSE}
module purge
module load compbio/homer

BEDS="" # populate BEDS variable according to needs
mkdir homer
for f in $BEDS; do
    cat $f |awk  'BEGIN{OFS="\t"}NR>1{print $1, $2, $3, $1"_"$2"_"$3"_"$4, "."}' \
    >homer/${f/bed/tsv} && \
    annotatePeaks.pl homer/${f/bed/tsv} hg38 \
      -annStats homer/${f/bed/annDefaultStats.txt} \
      2>/dev/null \
      >homer/${f/bed/ann.default.bed} &
done
```

# Set blacklist

```{r, cache=TRUE}
dar_ge_corr_blacklist <- c("PC_15194", "PC_22603", "PC_4786")
blacklist <- list( rna = c("BPH_337", "PC_4538", "PC_6102", "PC_6864", "CRPC_531", dar_ge_corr_blacklist), 
                   mirna = c("PC_6864", "CRPC_531", dar_ge_corr_blacklist))

```

# Load data

## RNA-seq

```{r, cache=TRUE}
load("R_data/RNAdds.RData")
load("R_data/RNA_differentiallyExpressed.RData")

dds <- rna_dds[,!colnames(rna_dds) %in% blacklist[["rna"]]]
rc_rna <- assay(dds, "norm.counts")
```

## sRNA-seq

```{r load_mirna}
load("R_data//sRNA_dds.RData")
dds <- srna_se_filter[,!colnames(srna_se_filter) %in% blacklist[["mirna"]]]
mirna_rcmat <- assay(dds, "norm.counts")

```

## Proteogenomic data

```{r}
sp_data <- fread("tables/swissprotdata_normalized.tsv")
rn <- sp_data[["uniprot"]]
prot <- as.matrix(sp_data[,-c("uniprot","descr","gn")])
rownames(prot) <- rn
```

## ATAC-seq

```{r, cache=TRUE, eval=FALSE}

DAR_MAT_PATH <- "dar/RCmat_mor_normed/turp_corrected.tsv.gz"

dar_rc_mat_norm <- fread(cmd = sprintf("zcat %s", DAR_MAT_PATH), sep ="\t", header = TRUE, showProgress = FALSE)
names(dar_rc_mat_norm)[1] <- "chrom"
dar_rc_mat_norm[,dar_id := paste(chrom, start, end, sep = "_")]
cols2keep <- !names(dar_rc_mat_norm) %in% c("chrom", "start", "end")
dar_rc_mat_norm <- dar_rc_mat_norm[, ..cols2keep]

# for plotting
rn <- dar_rc_mat_norm[["dar_id"]]
atac_rc <- as.data.frame(dar_rc_mat_norm[,-"dar_id",with=FALSE])
rownames(atac_rc) <- rn

# for other processing
names(dar_rc_mat_norm) <- gsub(pattern = "^(?:TURP|RP)_(.*)", replacement="\\1", x = names(dar_rc_mat_norm))
```

## Ensembl annotation

```{r, cache=TRUE}
source("/bmt-data/genomics//projects//atac_workdir/R/getAnnotationTable.R")
ann <- getAnnotationTable()
```

## TAD annotation

```{r}
# lncap_tad <- import("/bmt-data/genomics/reference/TAD/hg38.TADs/LNCAP_raw-merged_TADs.txt", format = "bed")
lncap_tad <- import("/bmt-data/genomics/projects/atac_workdir/dar_and_tad/lncap/lncap_tads.bed", format = "bed")
start(lncap_tad) <- start(lncap_tad) - 1

esc_tad <- import("/bmt-data/genomics/projects/atac_workdir/dar_and_tad/h1_esc/h1_esc_tads.bed", format = "bed")
start(esc_tad) <- start(esc_tad) - 1
```

## Uniprot

```{r}
if (!file.exists("R_data/ensembl2uniprot_mapping.Rdata")){
  uniprot <- fread(cmd = "zcat /bmt-data/genomics/reference/uniprot/HUMAN_9606_idmapping.dat.gz", header = FALSE)
  names(uniprot) <- c("uniprot_acc", "feature", "feature_id")
  uniprot <- uniprot[feature == "Ensembl"]
  uniprot[,feature := NULL]
  uniprot <- aggregate(. ~ feature_id, data = uniprot, FUN = function (x) paste(unique(x), collapse=";"))
  save(uniprot, file = "R_data/ensembl2uniprot_mapping.Rdata")
  # setkey(uniprot, "feature_id")
  # names(map_proteins) <- c("Nearest Ensembl", "protein_id")
} else {
  load("R_data/ensembl2uniprot_mapping.Rdata")
}
```

# Define samples list

```{r}
colsWhitelist <- !names(dar_rc_mat_norm) %in% dar_ge_corr_blacklist

all_samples_rna   <- setdiff(intersect(names(dar_rc_mat_norm), colnames(rc_rna)), blacklist[["rna"]])
all_samples_mirna <- setdiff(intersect(names(dar_rc_mat_norm), colnames(mirna_rcmat)), blacklist[["mirna"]])
all_samples_prot  <- intersect(names(dar_rc_mat_norm), colnames(prot))

samples <- list(all      = list(rna   = all_samples_rna,
                                mirna = all_samples_mirna,
                                prot  = all_samples_prot),
                PCvBPH   = list(rna   = grep(pattern = "^PC|BPH",   x = all_samples_rna,   value = TRUE), 
                                mirna = grep(pattern = "^PC|BPH",   x = all_samples_mirna, value = TRUE),
                                prot  = grep(pattern = "^PC|BPH",   x = all_samples_prot,  value = TRUE)),
                CRPCvPC  = list(rna   = grep(pattern = "^CRPC|PC",  x = all_samples_rna,   value = TRUE), 
                                mirna = grep(pattern = "^CRPC|PC",  x = all_samples_mirna, value = TRUE),
                                prot  = grep(pattern = "^CRPC|PC",  x = all_samples_prot,  value = TRUE)),
                CRPCvBPH = list(rna   = grep(pattern = "^CRPC|BPH", x = all_samples_rna,   value = TRUE), 
                                mirna = grep(pattern = "^CRPC|BPH", x = all_samples_mirna, value = TRUE),
                                prot  = grep(pattern = "^CRPC|BPH", x = all_samples_prot,  value = TRUE)))
```

# Compute correlation

## Define helper functions

```{r, cache=TRUE}
getDarTable <- function (dar_table_path) {
  t <- fread(dar_table_path)
  names(t)[1] <- "dar_id_orig"
  t[,dar_id:=gsub(pattern="(chr.*_[0-9]+_[0-9]+)(?:_.*)$", replacement = "\\1", x = dar_id_orig)]
  # t <- merge(t,map_proteins,by="Nearest Ensembl",all.x=TRUE)
  t
}

getDars <- function (dar_table_path) {
  dar_table <- getDarTable(dar_table_path)
  idx <- sort(dar_rc_mat_norm[dar_table, on=.(dar_id), which=TRUE, nomatch=0])
  rcmat <- dar_rc_mat_norm[idx, ..colsWhitelist]
  rid <- rcmat[,dar_id]
  rcmat[,dar_id:=NULL]
  m <- as.matrix(rcmat)
  rownames(m) <- rid
  counts <- c(
    dars_in_segment=nrow(fread(gsub(x=dar_table_path, pattern="10k_filter.e90ann.bed", replacement="tsv"))),
    dars_10kbpTSS=nrow(dar_table),
    unique_genes=nrow(unique(dar_table[,"Nearest Ensembl"])))
  dar_table <- merge(dar_table, uniprot, all.x = TRUE, by.x = "Nearest Ensembl", by.y = "feature_id")
  list(table=dar_table, mat=m, counts=counts)
}
```

```{r, eval = FALSE, cache=TRUE}
remove_olaps <- function (c) {
  suppressPackageStartupMessages(require(GenomicAlignments))

  split_names <- data.frame(
    chr=gsub(x=names(c),pattern="^(chr.*)_[0-9]+_[0-9]+(?:_.*)*$",replacement="\\1"),
    start=gsub(x=names(c),pattern="^chr.*_([0-9]+)_[0-9]+(?:_.*)*$",replacement="\\1"),
    end=gsub(x=names(c),pattern="^chr.*_[0-9]+_([0-9]+)(?:_.*)*$",replacement="\\1"),
    strand=rep("*",length(c)),
    cor=c)

  tryCatch({
    segments <- makeGRangesFromDataFrame(split_names,keep.extra.columns=TRUE)
    segments <- sort(segments)
  }, error = function (err) {
    write.csv(split_names,file="~/test/errDump.csv")
    cat("[remove_overlaps][ERR] ---> errDump.csv in ~/test <---\n")
    stop(err)
  })
  
  keep <- rep(FALSE, length(segments))
  pb <- txtProgressBar(min = 1, max = length(segments), initial = 0, char = "#", width = 60, style = 3, file = stderr())
  for(i in 1:length(segments)) {
    if(keep[i]) next
    
    candidate <- segments[i]
    olaps <- findOverlaps(segments, candidate)
    if (length(olaps) > 1) {
      h <- queryHits(olaps)
      if(any(keep[h])) next
      olapping_segs <- segments[h]
      if(all(is.na(olapping_segs$pearson))) next
      ii <- which.max(abs(olapping_segs$pearson))
    } else {
      ii <- i
    }
    keep[ii] <- TRUE
    setTxtProgressBar(pb,i)
  }
  close(pb)
  
  gres <- segments[keep]
  res <- gres$cor
  names(res) <- paste(seqnames(gres), start(gres), end(gres), sep="_")
  
  return(res)
}
```

## Init output count table

```{r}
countsTable <- data.frame(row.names=c("total_dars", "dars_10kbpTSS", "unique_genes"))
```

```{r}
target <- c(#"dar/turp_corrected_log2ratio_2/with_comparison_group_col/homer/combined_2_comparisons.ann.default.bed",
            #"dar/turp_corrected_log2ratio_2/with_comparison_group_col/homer/combined_3_comparisons.ann.default.bed",
            "dar/turp_corrected_log2ratio_2/with_comparison_group_col/homer/dars_pc_bph.ann.default.bed",
            "dar/turp_corrected_log2ratio_2/with_comparison_group_col/homer/dars_crpc_pc.ann.default.bed")
            #"dar/turp_corrected_log2ratio_2/with_comparison_group_col/homer/dars_crpc_bph.ann.default.bed")

corrList <- vector("list", length = length(target))
names(corrList) <- gsub(pattern = "_", replacement = "v", 
                        x = toupper(gsub(pattern = ".+_(.+_.+)\\.ann\\.default\\.bed", replacement = "\\1", x = basename(target))))

for (i in seq_along(target)) {
  comparison <- gsub(pattern = ".+_(.+_.+)\\.ann\\.default\\.bed", replacement = "\\1", x = basename(target[i]))
  comparison <- gsub("_","v",toupper(comparison))
  
  l <- getDars(target[i])
  corrList[[i]] <- doCorr(l$table, l$mat, as.matrix(rc_rna), mirna_rcmat, prot, samples[[comparison]])
}


```

## Remove overlaps

```{r, eval = FALSE}
pearson <- list(
  PCvBPH  = corrList$PCvBPH$all_cor,
  CRPCvPC = corrList$CRPCvPC$all_cor
  # CRPCvBPH = corrList$CRPCvBPH$all_cor,
  # combo2comparison = corrList$comp2$all_cor,
  # combo3comparison = corrList$comp3$all_cor
)
```

```{r, eval = FALSE}
pearson_no_olaps <- list(
  PCvBPH  = remove_olaps(corrList$PCvBPH$all_cor),
  CRPCvPC = remove_olaps(corrList$CRPCvPC$all_cor)
  # CRPCvBPH = remove_olaps(corrList$CRPCvBPH$all_cor),
  # combo2comparison = remove_olaps(corrList$comp2$all_cor),
  # combo3comparison = remove_olaps(corrList$comp3$all_cor)
)
```

```{r, eval = FALSE}
spearman <- list(
    PCvBPH  = corrList$PCvBPH$all_cor_spearman,
    CRPCvPC = corrList$CRPCvPC$all_cor_spearman
    # CRPCvBPH = corrList$CRPCvBPH$all_cor_spearman,
    # combo2comparison = corrList$comp2$all_cor_spearman,
    # combo3comparison = corrList$comp3$all_cor_spearman
)
```

```{r, eval = FALSE}
spearman_no_overlap <- list(
  PCvBPH  = remove_olaps(corrList$PCvBPH$all_cor_spearman),
  CRPCvPC = remove_olaps(corrList$CRPCvPC$all_cor_spearman)
  # CRPCvBPH = remove_olaps(corrList$CRPCvBPH$all_cor_spearman),
  # combo2comparison = remove_olaps(corrList$comp2$all_cor_spearman),
  # combo3comparison = remove_olaps(corrList$comp3$all_cor_spearman)
)
```

### Plot correlation coefficients distributions

```{r, eval = FALSE}
doViolin <- function (L) {
  require("ggplot2")
  require("reshape")

  df <- melt(L)
  df$contrast <- factor(gsub(x = df$L1, pattern = ".*-(.*)", replacement = "\\1"))
  df$segment <- factor(gsub(x = df$L1, pattern = "(.*)_.*", replacement = "\\1"))
  df$mat <- factor(gsub(x = df$L1, pattern = ".*_(.*-mat)-.*", replacement = "\\1"))

  p <- ggplot(df, aes(contrast, value, fill=contrast)) +
    geom_violin(draw_quantiles = c(.25, .5, .75), trim = FALSE, scale = "count") +
    #facet_grid(mat ~ segment) +
    theme_bw() +
    theme(
      text = element_text(family = "sans")) +
      #axis.text.y = element_blank(),
      #axis.title.y = element_blank()) +
    ylab("Correlation coefficient")
    #ggtitle("u = uncorrected, c = background corrected, j = Juha's correction method") +
    #coord_flip()
    #geom_jitter()

  return(p)
}
```

```{r, eval = FALSE}
# PEARSON VIOLINS
p1 <- doViolin(pearson) + ggtitle("Pearson correlation - all")
p3 <- doViolin(pearson_no_olaps) + ggtitle("Pearson correlation - overlapping DARs removed")
ppear <- marrangeGrob(grobs = list(p1, p3), nrow = 1, ncol = 1)
ggsave(filename = "/bmt-data/genomics/projects/atac_workdir/pictures/correlation_pearson_violins_turp_v2.pdf", 
       plot = ppear, device = "pdf", width = 15, height = 8, units = "in")
```

```{r, eval = FALSE}
# SPEARMAN VIOLINS
p4 <- doViolin(spearman) + ggtitle("Spearman correlation - all")
p5 <- doViolin(spearman_no_overlap) + ggtitle("Spearman correlation - overlapping DARs removed")
pspear <-  marrangeGrob(grobs = list(p4, p5), nrow = 1, ncol = 1)
ggsave("/bmt-data/genomics/projects/atac_workdir/pictures/correlation_spearman_violins_turp_v2.pdf", 
       plot = pspear, device = "pdf", width = 15, height = 8)
```

```{r, eval = FALSE}
# COUNTS TABLE
ggsave("/bmt-data/genomics/projects/atac_workdir/pictures/correlation_tables_turp.pdf", 
       plot = tableGrob(countsTable), device = "pdf", width = 15, height = 8)
```

# Merge Homer annotation, correlation results, RNA-seq results, ATAC-seq results and LnCAP TAD annotation

## Define helper

```{r}
extendTable <- function(t, deg_table, rna_table) {
    ensgids <- t[, "Nearest Ensembl"]

    t$rna_de <- ensgids %in% rownames(deg_table)
    t$rna_log2_ratio <- rna_table[match(ensgids, rownames(rna_table)), "log2FoldChange"]
    t$rna_padj <- rna_table[match(ensgids, rownames(rna_table)), "padj"]
    t$rna_abs_med_diff <- rna_table[match(ensgids, rownames(rna_table)), "diff"]
    
    t$rna_log2_ratio <- round(t$rna_log2_ratio, digits=2)
    t$rna_padj <- round(t$rna_padj, digits=6)
    t$rna_abs_med_diff <- round(t$rna_abs_med_diff, digits=2)
    
    t
}

# extend2ComparisonTable <- function (t) {
#     t1 <- extendTable(t, de_PCvBPH, rna_PCvBPH)
#     t2 <- extendTable(t, de_CRPCvPC, rna_CRPCvPC)
#     toMerge <- c("rna_de", "rna_log2_ratio", "rna_padj", "rna_abs_med_diff")
#    
#     tf <- merge(t1[,toMerge], t2[,toMerge], by=0, suffixes = c(".PCvBPH", ".CRPCvPC"))[-1]
#     cbind(t,tf)
# }
# 
# extend3ComparisonTable <- function (t) {
#   t1 <- extend2ComparisonTable(t)
#   t2 <- extendTable(t, de_CRPCvBPH, rna_CRPCvBPH)
#   toMerge <- c("rna_de", "rna_log2_ratio", "rna_padj", "rna_abs_med_diff")
#   tm <- t2[,toMerge]
#   colnames(tm) <- paste(colnames(tm), "CRPCvBPH", sep=".")
#   
#   cbind(t1,tm)
# }

getDarsRawBed <- function (path) {
  require(data.table)
  dt <- fread(path)
  dt[,dar_id:=paste(chrom,start,end,comparison_group,sep="_")]
  rn <- dt[,dar_id]
  df <- as.data.frame(dt[,dar_id:=NULL])
  rownames(df) <- rn
  toFormat <- c("med_pc", "med_bph", "med_crpc", "abs_med_diff", "log2_ratio", "u_stat")
  for(i in 1:length(toFormat)) {
    if(toFormat[i]%in%colnames(df)){
      df[,toFormat[i]] <- format(df[,toFormat[i]], digits=2)
    }
  }
  df$p_value <- format(round(df$p_value, digits=6), scientific=FALSE)
  df
}

mergeWithDarBed <- function(df, bed) {
  b <- bed[,-which(colnames(bed)%in%c("chrom","start","end","comparison_group"))]
  colnames(b) <- paste("atac", colnames(b), sep="_")
  df <- merge(df, b, by=0)
  rownames(df)<-df[,1]
  df[,-1]
}

mergeWithHomerAnnot <- function (df, homer) {
  df <- df[,-which(colnames(df)%in%c("chr","start","end","Nearest.Ensembl","Distance.to.TSS"))]
  
  df$pearson <- round(df$pearson, digits=2)
  df$spearman <- round(df$spearman, digits=2)
  df$pearson_protein <- round(df$pearson_protein, digits=2)
  df$spearman_protein <- round(df$spearman_protein, digits=2)
  
  merged <- merge(homer,df,by=0)
  rn <- merged[,"Row.names"]
  rownames(merged) <- rn
  merged[,-1]
}

mergeWithTad <- function (dar.tbl, tad, colname = NULL)    {
  require(GenomicAlignments)
  dar.tbl$dar_id <- rownames(dar.tbl)
  dar.tbl <- as.data.table(dar.tbl)
  dar.gr <- makeGRangesFromDataFrame(as.data.frame(dar.tbl), 
                                     keep.extra.columns = TRUE, 
                                     starts.in.df.are.0based = FALSE)
  olaps <- findOverlaps(dar.gr, tad)
  dt <- data.table(dar_id = dar.tbl[["dar_id"]][queryHits(olaps)],
                   tad = as.character(tad[subjectHits(olaps)]))
  dt[, tad := paste(tad, collapse = ";"), by=dar_id]
  dt <- unique(dt, by = "dar_id")
  if (!is.null(colname)) names(dt)[2] <- colname
  merged <- merge(dar.tbl,dt, by="dar_id", all.x=TRUE)
  rn <- merged[["dar_id"]]
  merged <- as.data.frame(merged)
  rownames(merged) <- rn
  merged
}

getHomerAnnotated <- function (path) {
  dt <- fread(path)
  names(dt)[1] <- "dar_id"
  rn <- dt[,dar_id]
  df <- as.data.frame(dt)
  rownames(df)<-rn
  df[,-1]
}

```

## Set output dir

```{r}
output_folder <- "tables/correlation_turp/dar_v2"
if(!dir.exists(output_folder)) dir.create(output_folder)
outfile <- "turp_corrected.csv" # suffix
```

## Join and dump all tables

```{r}
library(openxlsx)

target_bed <- c("dar/turp_corrected_log2ratio_2/with_comparison_group_col/dars_pc_bph.bed", 
                "dar/turp_corrected_log2ratio_2/with_comparison_group_col/dars_crpc_pc.bed"
                # "dar/turp_corrected_log2ratio_2/with_comparison_group_col/dars_crpc_bph.bed",
                # "dar/turp_corrected_log2ratio_2/with_comparison_group_col/combined_2_comparisons.bed",
                # "dar/turp_corrected_log2ratio_2/with_comparison_group_col/combined_3_comparisons.bed"
                )

target_de  <- list(de_PCvBPH,  
                   de_CRPCvPC
                   # de_CRPCvBPH
                   )

target_rna <- list(rna_PCvBPH, 
                   rna_CRPCvPC
                   # rna_CRPCvBPH
                   )

results <- vector("list", length = length(corrList))
names(results) <- names(corrList)
for (i in seq_along(target)) {
  
  bed         <- getDarsRawBed(target_bed[i])
  homer_annot <- getHomerAnnotated(target[i])
  
  t <- mergeWithHomerAnnot(corrList[[i]][["table"]],homer_annot)
  t <- extendTable(t, target_de[[i]], target_rna[[i]])
  t <- mergeWithDarBed(t, bed)
  t <- mergeWithTad(t, lncap_tad, "TAD LNCaP")
  t <- mergeWithTad(t, esc_tad,   "TAD H1 ESC")
  results[[i]] <- t
}

write.xlsx(results, file = file.path(output_folder, "dar_correlation_homer_closest_gene.xlsx"))
```

## Define helper

```{r, eval = FALSE}
doScatterplot <- function (t) {
  
  columns <- intersect(colnames(rc_rna),gsub(x=colnames(atac_rc), pattern="^(?:TURP|RP)_(.*)$", replacement = "\\1"))

  atac_rc <- atac_rc[,match(columns, gsub(x=colnames(atac_rc), pattern="^(?:TURP|RP)_(.*)$", replacement = "\\1"))]
  rc_rna <- rc_rna[,match(columns,colnames(rc_rna))]
  
  #stopifnot(all(colnames(atac_rc)==colnames(rc_rna)))
  
  grobs <- list()
  for(i in 1:nrow(t)) {
    dar <- gsub(x = rownames(t)[i], pattern = "^(chr.*_[0-9]+_[0-9]+)_.*", replacement = "\\1")
    g   <-  as.character(t[i,"Nearest Ensembl"])
    distance <- t[i, "Distance to TSS"]
    gn  <- ann$gene_name[ann$gene_id == g & ann$type == "gene"]
    
    if(g %in% rownames(rc_rna)){
      tryCatch({
        df <- data.frame(
          atac=unname(t(atac_rc[dar,])),
          rna=rc_rna[g,],
          sampling.method=factor(gsub(pattern="^(TURP|RP)_.*", replacement="\\1", x=colnames(atac_rc))),
          type=factor(gsub(x=colnames(atac_rc),pattern="^.*_(BPH|PC|CRPC)_[0-9]+",replacement = "\\1")),
          RNA.protocol=colData(rna_dds)[match(gsub(pattern="^(?:TURP|RP)_(.*)",replacement="\\1",x=colnames(atac_rc)),colnames(rna_dds),nomatch=0),"RNA.Isolation.method"])
      }, error = function (err) {
        print(dar)
        print(g)
        print(atac_rc[dar,])
        print(rc_rna[g,])
        print(factor(gsub(pattern="^(TURP|RP)_.*", replacement="\\1", x=colnames(atac_rc))))
        print(factor(gsub(x=colnames(atac_rc),pattern="^.*_(BPH|PC|CRPC)_[0-9]+",replacement = "\\1")))
        print(colData(rna_dds)[match(gsub(pattern="^(?:TURP|RP)_(.*)",replacement="\\1",x=colnames(atac_rc)),colnames(rna_dds),nomatch=0),"RNA.Isolation.method"])
        stop(err)
      })
      
      map_type <- c(BPH=21,PC=22,CRPC=23)
      df$typeNum <- map_type[match(df$type, names(map_type))]
      
      df1 <- df
      df1$atac[df1$atac > ceiling(xlim[2])] <- xlim[2]
      df1$rna[df1$rna > ceiling(ylim[2])] <- ylim[2]
      
      p1 <- ggplot(df1,aes(atac, rna,
                           shape = factor(typeNum),
                           color = factor(sampling.method),
                           fill = factor(RNA.protocol))) +
        scale_fill_manual(guide = "none", values = c("lightyellow", "pink")) +
        scale_shape_manual(guide = "none", values = c(21, 22, 23), labels = names(map_type)) +
        scale_color_manual(guide = "none", values = c("red", "blue")) +
        scale_x_continuous(breaks = seq(0, ceiling(xlim[2]), by = 1), limits = xlim) +
        scale_y_continuous(breaks = seq(0, ceiling(ylim[2]), by = 1), limits = ylim) +
        geom_point(size = 1.5) +
        xlab("ATAC RC - 0:90%") +
        ylab("RNA RC (log2) - 0:90%") +
        theme_bw()
      
      p2 <- ggplot(df,aes(atac, rna,
                          shape = factor(typeNum),
                          color = factor(sampling.method),
                          fill = factor(RNA.protocol))) +
        scale_fill_manual(values = c("lightyellow", "pink"), name="RNA extraction\nmethod") +
        scale_shape_manual(values = c(21, 22, 23), labels = names(map_type), name="Sample type") +
        scale_color_manual(values = c("red", "blue"), name="Sample collection\nmethod") +
        geom_point(size = 1.5) +
        xlab("ATAC RC") +
        ylab("RNA RC (log2)") +
        theme_bw() +
        theme(legend.position="bottom") +
        guides(
          fill=guide_legend(nrow=2, byrow=TRUE, override.aes = list(shape = 21)),
          color=guide_legend(nrow=2, byrow=TRUE),
          shape=guide_legend(nrow=2, byrow=TRUE))
      
      legend <- get_legend(p2)
      p2 <- p2 + theme(legend.position="none")
      
      plots <- plot_grid(p1, p2, ncol = 2, rel_widths = c(1, 1))
      
      title <- ggdraw() + draw_label(sprintf("%s - %s; distance to TSS: %d\nPearson = %.3f; Spearman = %.3f", dar, gn, distance, as.numeric(t[i,"cor"]),as.numeric(t[i,"cor_spearman"])), fontface = 'bold')
      
      grobs[[i]] <- plot_grid(title, plots, legend, ncol = 1, nrow = 3, rel_heights = c(0.2, 1, 0.2))
    } else {
      next
    }
  }
  grobs
  
}
```

## Set output path

```{r, eval = FALSE}
output_folder <- "pictures/scatterplot_correlation_turp_v3"
#output_folder <- "~/projects/atacseq_tamperepc/pictures/scatterTest"
if(!dir.exists(output_folder)) dir.create(output_folder,recursive = TRUE)
```

## Set axis limits (right panel)

```{r, eval = FALSE}
xlim <- c(0, quantile(as.matrix(dar_rc_mat_norm[,-"dar_id",with=FALSE]), .90))
ylim <- c(0, quantile(rc_rna, .90))
```

## Do plots

```{r}
for (i in seq_along(results)) {
  
  t <- subset(results[[i]], subset = abs(as.numeric(pearson)) > 0.2 | abs(as.numeric(spearman)) > 0.2)
  o <- order(t[,"Distance to TSS"], decreasing = TRUE)
  g <- suppressWarnings(doScatterplot(t[o,]))
  
  outfile <- sprintf("%s_turp_corrected.pdf", names(results)[i])
  ggsave(file.path(output_folder,outfile),plot=marrangeGrob(grobs=g,nrow=2,ncol=1),height=8,width=8)
  
}
```
