---
title: 'Tampere PC: correlation of gene expression and ATAC peaks in promoter region'
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Francesco Tabaro"
output:
  pdf_document:
  toc: true
toc_depth: 2
number_sections: true  
df_print: tibble
fig_caption: true
html_document:
  toc: true
toc_depth: 2
number_sections: true
fig_caption: true
df_print: paged
toc_float:
  collapsed: false
smooth_scroll: false
---

```{r setup}
knitr::opts_knit$set(base.dir = "/bmt-data/genomics/projects/atac_workdir/")

library(data.table)
library(DESeq2)

NCORES <- floor(parallel::detectCores() * 0.5)
setDTthreads(NCORES)
```

# Load data

## RNA-seq

```{r}
load("/bmt-data/genomics/projects/atac_workdir/R_data/RNAdds.RData")
rna_rc <- assay(rna_dds, "norm.counts")
```

## sRNA-seq

```{r}
load("/bmt-data/genomics/projects/atac_workdir/R_data/sRNA_dds.RData")
srna_rc <- assay(srna_se_filter, "norm.counts")
```

## Promoter ranges

```{r}
gr <- granges(c(rowRanges(rna_dds), rowRanges(srna_se_filter)))
proms <- promoters(gr, upstream = 1000, downstream = 100)
```


## ATAC-seq signal 

```{r}
PEAKS_RC_PATH <- "/bmt-data/genomics/projects/atac_workdir/peak_tss_dar_quantifications/peak_quantification.tsv.gz"
atac_rc_peaks <- fread(cmd = sprintf("zcat %s", PEAKS_RC_PATH))
atac_rc_peaks[,`:=`(peak_id = paste(chrom, start, end, sep = "_"), chrom=NULL, start=NULL, end=NULL)]
```

## Peaks ranges

```{r}
atac_peaks <- fread("/bmt-data/genomics/projects/atac_workdir/heterogeneity/peaks_and_their_corresponding_samples.tsv")
names(atac_peaks) <- c("chrom", "start", "end", "samples")
atac_peaks[,peak_id := paste(chrom, start, end, sep = "_")]
atac_peaks <- makeGRangesFromDataFrame(as.data.frame(atac_peaks), keep.extra.columns = TRUE, starts.in.df.are.0based = TRUE)
names(atac_peaks) <- atac_peaks$peak_id
```

# Fix samples

```{r}
deg_blacklist <- c("BPH_337", "PC_4538", "PC_6102", "PC_6864", "CRPC_531")
peaks_ge_corr_blacklist <- c(deg_blacklist, "PC_15194", "PC_22603", "PC_4786")

samples <- Reduce(intersect, list(colnames(rna_dds), colnames(atac_rc_peaks), colnames(srna_se_filter))) # 28 samples

```

# Build unified RNA matrix

```{r}
rna_rc <- rbind(rna_rc[,samples], srna_rc[,samples])
```

# Find peaks in promoters

```{r}
olaps <- findOverlaps(proms, atac_peaks)
proms_with_olap <- names(proms)[queryHits(olaps)]
peaks_with_olap <- names(atac_peaks)[subjectHits(olaps)]
```

# Compute correlation

```{r}
res <- do.call(rbind, mclapply(seq_along(olaps), function(i) {
# for(i in seq_along(olaps)) {
  gid <- proms_with_olap[i]
  pid <- peaks_with_olap[i]

  rna <- rna_rc[gid, samples]
  
  atac <- atac_rc_peaks[peak_id == pid, -"peak_id", with = FALSE]
  atac <- atac[,..samples]
  atac <- setNames(as.numeric(atac), names(atac))
  
  stopifnot(all(names(rna) == names(atac)))
  
  pearson  <- round(cor(rna,atac,method = "pearson"),2)
  spearman <- round(cor(rna,atac,method = "spearman"),2)
  
  data.frame(gene_id = gid, peak_id=pid, pearson = pearson, spearman = spearman, stringsAsFactors = FALSE)
  # print(c(gid,pid))
}, mc.cores = NCORES))
```

```{r}
print(dim(res))
```


# Extend table

## Load Ensembl annotation

```{r}
source("/bmt-data/genomics/projects/atac_workdir/R/getAnnotationTable.R")
ann <- getAnnotationTable()
ann.gene <- as.data.frame(subset(ann, type == "gene"))
ann.gene <- ann.gene[,sapply(ann.gene, function(col) !all(is.na(col)))]
```

## Convert miRNA names to Ensembl gene IDs

```{r}
mir <- grep("hsa", res$gene_id, value = TRUE, ignore.case = TRUE)
pat <- gsub(pattern = "hsa-(?:mir|let)-([a-z0-9]+(?:-[0-9])?)(?:-[0-9a-z]+)?$", replacement = "mir\\1$", x = mir, ignore.case = TRUE)
idx <- sapply(pat, function(x){ 
  i <-  grep(pattern = x, x = ann.gene$gene_name, ignore.case = TRUE)
  ifelse(length(i) == 0, NA_integer_, i[1])
})

mir_map <- data.frame(mir=mir,ensg=ann.gene$gene_id[idx], stringsAsFactors = FALSE)
res$gene_id_map <- NA
res$gene_id_map[grep("ensg", res$gene_id, ignore.case = TRUE)] <- res$gene_id[grep("ensg", res$gene_id, ignore.case = TRUE)]
res$gene_id_map[grep("hsa", res$gene_id, ignore.case=TRUE)] <- mir_map[match(res$gene_id[grep("hsa", res$gene_id, ignore.case=TRUE)], mir_map$mir), "ensg"]
```

```{r}
print(dim(res))
```


## Join correlation table to annotation

```{r}
res <- merge(res,ann.gene, by.x="gene_id_map", by.y="gene_id", all.x = TRUE)
print(dim(res))
```

## Join with de status

```{r}
load("/bmt-data/genomics/projects/atac_workdir/R_data/RNA_differentiallyExpressed.RData")
res$PCvBPH_de  <- res$gene_id_map %in% rownames(de_PCvBPH)
res$PCvBPH_lfc <- round(rna_PCvBPH$log2FoldChange[match(res$gene_id_map, rownames(rna_PCvBPH))],2)
res$CRPCvPC_de <- res$gene_id_map %in% rownames(de_CRPCvPC)
res$CRPCvPC_lfc <- round(rna_CRPCvPC$log2FoldChange[match(res$gene_id_map, rownames(rna_CRPCvPC))],2)
print(dim(res))
```

## Add peaks sample list

```{r}
tmp <- data.frame(peak_id=names(atac_peaks),peak_samples=atac_peaks$samples)
res <- merge(res,tmp,by="peak_id", all.x=TRUE)
print(dim(res))
```

## Add GTRD TFBS annotation

```{r}
mergeGtrd <- function (dt) {
  df <- data.frame(chr   = gsub(pattern = "(chr.*)_[0-9]+_[0-9]+", replacement = "\\1", x = dt[["peak_id"]]),
                   start = gsub(pattern = "chr.*_([0-9]+)_[0-9]+", replacement = "\\1", x = dt[["peak_id"]]),
                   end   = gsub(pattern = "chr.*_[0-9]+_([0-9]+)", replacement = "\\1", x = dt[["peak_id"]]),
                   name  = dt[["peak_id"]])
  gr    <- makeGRangesFromDataFrame(df, starts.in.df.are.0based = FALSE, keep.extra.columns = TRUE)
  olaps <- findOverlaps(gtrd, gr, type = "within")
  tf_df <- data.frame(peak_id = gr$name[subjectHits(olaps)], gtrd_tf = gtrd$name[queryHits(olaps)])
  tf_df <- aggregate(. ~ peak_id, data = tf_df, FUN = function(x) paste(as.character(unique(tf_df[x, "gtrd_tf"])), collapse = ", "))
  merge(dt, tf_df, by = "peak_id", all.x = TRUE)
}

gtrd <- import("/bmt-data/genomics/reference/gtrd/gtrd_meta_clusters_prostate_w_names.bed")

res <- mergeGtrd(as.data.table(res))
print(dim(res))
```

# Save

```{r}
library(xlsx)
write.xlsx(res, file = "/bmt-data/genomics/projects/atac_workdir/tables/correlation_turp/correlation_peaks.xlsx")
```

# Make histograms

```{r}
pdf("/bmt-data/genomics/projects/atac_workdir/pictures/correlation_turp/correlation_peaks.pdf")
par(mfrow=c(2,1), oma = c(0,0,3,0))
hist(res$pearson, breaks = "fd", col = "grey90", main = "Pearson", xlab = sprintf("Pearson correlation coefficient - median = %.2f", median(res$pearson)))
abline(v=median(res$pearson), col = "dodgerblue", lwd = 2)

hist(res$spearman, breaks = "fd", col = "grey90", main = "Spearman", xlab = sprintf("Spearman correlation coefficient - median = %.2f", median(res$spearman)))
abline(v=median(res$spearman), col = "dodgerblue", lwd = 2)

mtext("Correlation between RNA-seq signal of genes\nand ATAC-seq signal of peaks in promoter region (-1kbp to 100bp around TSS)", outer = TRUE, line = 1)
dev.off()
```


